{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trophic model for gut\n",
    "## Consumer-Resource model (CRM) with Trophic levels\n",
    "The human gut microal community is complex because of 3 reason: (1) many microbial species (~570); (2) many metabolites involoved (~244); and (3) many microbe-metabolite interactions/links (>4400). Of all types of microbe-metabolite interactions/links, the cross-feeding makes the system more complicated to interpret. Previously, a literature-curated interspecies network of the human gut microbiota, called [NJS16](https://www.nature.com/articles/ncomms15393) is reported. This is an extensive data resource composed of âˆ¼570 microbial species and 3 human cell types metabolically interacting through >4,400 small-molecule transport and macromolecule degradation events.\n",
    "\n",
    "Here, we devoted to build a Consumer-Resource model (CRM) with trophic levels posted in the paper \"[a evidence for a multi-level trophic organization of the human gut microbiome](https://www.biorxiv.org/content/10.1101/603365v1.full-text)\". The model is devoted to study the complex microbe-metabolite network in human guts.\n",
    "\n",
    "In concise, in the mode, a trophic level is considered as one round of carbon processing and is composed of two process: resource allocation to microbes and the following resource/byproduct generation by microbes. The simulation is stopped after several trophic levels because the residence time of resource in the human gut is assumed to be finite.\n",
    "\n",
    "Previously, the code for the model is written in Matlab. Here a Python version of the model is proposed, providing further convenience for Machine Learning model to learn uptake rates and byproduct generation rates of all microbial species to different metabolites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "########### Self-customized setting\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy as sp\n",
    "import sklearn\n",
    "import math\n",
    "from numba import jit, float64\n",
    "\n",
    "#%% Plot Tong's default setting\n",
    "SMALL_SIZE = 15\n",
    "MEDIUM_SIZE = 15\n",
    "BIGGER_SIZE = 15\n",
    "\n",
    "plt.rc('font', size=SMALL_SIZE)          # controls default text sizes\n",
    "plt.rc('axes', titlesize=SMALL_SIZE)     # fontsize of the axes title\n",
    "plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "plt.rc('xtick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('ytick', labelsize=SMALL_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('legend', fontsize=SMALL_SIZE)    # legend fontsize\n",
    "plt.rc('figure', titlesize=BIGGER_SIZE)  # fontsize of the figure title\n",
    "plt.rcParams[\"font.weight\"] = \"bold\"\n",
    "plt.rcParams[\"axes.labelweight\"] = \"bold\"\n",
    "plt.rcParams['lines.linewidth'] = 1.5\n",
    "plt.rcParams['axes.linewidth'] = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "########### import the pickled file containing all processed data which are useful for simulations (the processing is\n",
    "########### done in \"Trophic_model_for_gut_data_processing.ipynb\")\n",
    "import pickle\n",
    "pickle_in = open(\"Chia_network.pickle\",\"rb\")\n",
    "net, i_selfish, i_intake, names = pickle.load(pickle_in)\n",
    "\n",
    "pickle_in = open(\"Thai_data.pickle\",\"rb\")\n",
    "thai_metagenome_ID, thai_metagenome, thai_metabolome_ID, thai_metabolome = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimization terminated successfully.    (Exit mode 0)\n",
      "            Current function value: 3.3082972161368787\n",
      "            Iterations: 14\n",
      "            Function evaluations: 310\n",
      "            Gradient evaluations: 14\n",
      "     fun: 3.3082972161368787\n",
      "     jac: array([ 0.00000000e+00, -6.15325570e-03, -2.90572643e-05,  2.35739350e-03,\n",
      "        8.16664100e-03,  1.50786638e-02,  1.46860778e-02, -4.07725573e-04,\n",
      "        1.19361281e-03,  1.25773847e-02, -2.30303407e-03, -2.99483538e-04,\n",
      "        2.42887139e-02,  5.35169244e-02, -1.20905936e-02,  8.37150216e-03,\n",
      "        9.80427861e-02,  2.01314688e-04,  4.41354513e-03,  1.83877051e-02])\n",
      " message: 'Optimization terminated successfully.'\n",
      "    nfev: 310\n",
      "     nit: 14\n",
      "    njev: 14\n",
      "  status: 0\n",
      " success: True\n",
      "       x: array([5.00000000e-02, 2.82396174e-02, 1.49672925e+00, 9.00404585e-01,\n",
      "       2.66745755e-02, 1.52703370e-16, 2.49504760e-17, 4.09981488e-02,\n",
      "       1.58968702e-02, 0.00000000e+00, 1.44357890e-01, 1.43499947e+00,\n",
      "       2.88696654e-17, 1.82546248e-16, 7.70360787e-01, 6.09055403e-01,\n",
      "       0.00000000e+00, 1.96865045e-02, 5.99137613e-01, 2.74302861e-17])\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8070143454797384, 1.233742065563012e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9261608455145216, 7.237895745818847e-07)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAS0AAAEmCAYAAAA3JMYtAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO2deZhcVZn/P28WSJqEAIYduhvCLjCiAgIqcRQCiA4wiA6BAQQiIJKIywwJEIIGdH6yuBA1GSVKt4C4sIgEZMmwgwFlJ0GgE8JmEraGbCT9/v44p9K3b9dyqvreqrpV7+d5znPvPffcc9663fWts75HVBXDMIysMKjWBhiGYZSDiZZhGJnCRMswjExhomUYRqYw0TIMI1OYaBmGkSlMtFJCRC4QERWR5wrc/4e/f0GZ+e5T7jNGcyIis0VkXq3tSBoTrXRZCWwnIh+NRorI3kCbv18u+wBTE7DNaHy+A5xYayOSxkQrXd4D7gS+FIv/ko9/r+oWGQ2PiAwHUNXnVfXJWtuTNCZa6XMNcIyICIA/HuPj+yEiHxeR/xOR5SKyTERmichIf+9E4Mf+XH2Y6693EZFrROQl/+xTIjJJRAbF8t9TRO4XkZU+zWEiMk9EZofakbPFl7+HiPxFRN4TkWdF5Kg8n+lMEXlORFb5ZvHXY/cvEJGlIrKvt2WFiNwrItuJyGYicr2IvCsiz4jIv+bJ/xT/WVaJyEIR+XbJv4p77lQRecK/i9dF5HciMipy/xh/f5V/r9NFZEied/BhEZnr39Xf/fUGInKliLwtIi+IyH/Eyp7ry5sgIl3+M98sIlvH0n3P2/CuiCwWkU4R2SKWpktELhGR80RkMfCOj+/TPBSRjUTkf0XkFf+ZF4nIrFhe/yoiD0XeyQwRGRG5P9Z/5rEicp236wUROSPknSeCqlpIIQAXAEuBjYBVwCd8/CdxzcKN/P0LIs8c4NNeCxwGHA+8DPzO398U+AGgwMd82M3f+zQwDfgcMBaYBLwNnBPJvwV4FfgbcCTwH8B84J/A7FA7fJoTvR1PAF8DDgZuAlYD20TSnerTXeLTXAz0AP8de1fLgceA8cARwCLgXuAO4Jv+2duBZUBL5NlvAe8D04GDgP/2tp9Z4u9zrrfjJ8AhwFHAL4Ct/f2Dvd2/8ve/7fP9WYF38BXgUP8ZXsD9KOVsutrbGH0vc/07fcKXfSzwEvDXmJ2/9H+nA4GjgQeAp4HBkTRd/u96O/B54CgfPxuYF8vrWeCLPr/jgJmR+7v5v9/NwGeB04C3gDmRNGP9Z37Ov8ODfL4K7FOV71atv9yNGvwXcak/vwG4wp/PAK7353HRuge4K5bPv/p/iN399ZmAlihbgCHAZOCFSPxX/T/l1pG4fXz+s8u0I/eF/XIkzQeANcBp/nqQ/2JeGctrBk5Qh0XelQIHRtKc4ePOj8Tt5uMO9dcbAu8CU2P5Xwi8Fv1ix+5vhBPJS4u8wwfzvINvA2vx4hN5BydE0hzm434ZiRuFE63TI3FzfVxbJO4A/+whBWwaDGzt03wyEt+FE61hsfSz6StaTwJfK/KZr8GJUVQQj/Hl7eevx/rrCyNphgJLgO9V47tlzcPqcA1wtIisj/u17Nc0FJEWYD/gtyIyJBdwtY33gY8UK0BEhonINBH5B65GkKt9bBdp0uwNPKKqL+eeU9WHgdcHYMdtkbyW4Wpt2/iobYCtgOtiz1yLE5w9InGrcWKZ4x/+eGeeuFwTaj9gA+C6mK13AptH7IizHzAcuDLfTREZDHy4gN2D/PNR7ihmt6q+jftS92n6AY+q6sJIuvtw72+fiC2H+ub827gfhMX+1k5xG1S11MDO34FvicgZIhJ/Hl/uH1V1bSTu977cj8fSRv/u7+PErtD7ThQTrepwIzACJyIb4JpRcTbG/ZLOwIlDLqzC/ZJtW6KM7+OaUTNxv/Z7A9/194b54xa4L0+caFy5drwVu14dKW9Lf3w9liZ3vUkkrltVe2L59MlfVXNxufxH++NTMVvv8vGF3tkH/PHVAvdH4z5riN19bMxndyR+WCzun3nK/if+vYkbZb4RJ1TH48TyYz5dPK+4rfk4E7geOB+Y7/sZo4NEW8bz8QK2jOKfGfJ/vlQYUjqJMVBU9T0R+RPwdeA6Vc03avgWrtp9AfDnPPdfKVHMF4Afq+r/5CJE5LOxNK8BO+d5dtME7YiSE4XNYvGb++MbZeSVj9zzh5P/Szu/wHPL/HFLXBM9zlKc+KVld454/rm43Hs7EveD8kXNtftF2grkVdLHlKq+BZwFnCUie+Kau50i8riqPu3L7WOTr3V+gOQ+84Cxmlb1+CmuhvWzfDe9kD0I7Kyq8/KEnFisBtccjGUxHFcbwt8fTP+pFn8FPhodoRKRfej9MpZjRwiLcSL3hVj8MbgRrifKyCsfDwArgK0K2Npd4rkT8t30tYtHCtjd459Pgg+LSGvuQkQOwInGwz5qOPB+TrA845MoWFUfxw1iDAJ28dEPAUf6/50cR+EqN/cmUW4SWE2rSqjqXFznazG+DdwhIj3A74BuoBU3kjNFVRfgRn8AJorIncA7qjof+AvwVd+n9Qau0339WP5X4kZ8/iQi03Bfimm4X/No0yzEjpDP3CNu9v7PRWSZt/FA4HRgckAfTKn83/L5/9DXQO7GfQl3Aj6lqkcWee47wHQRWQ9Xo1zff75pvs9vKnCriFyJ64PcAzdZc5aqLs6XbwX8E/e3uADXtPo+rp9rjr//F2CSiFyO+8HbHzfiVxEici/wR1yHvOJGdt+jVyS/ixtZvl5Eforro/o+cKuqJiXUA6cavf3NGIiMHhZJ02f00MftC8zB1UTeww1vXwqM8vcF+B9cDaYHmOvjN8f9Q76Dayr9D73TDUZE8v8X4H5crWw+bnrBAuDyMu04MZ63j+8CfhCLOxPXQb0aNx3g66XeFb2jVLvH4pXYdAbcF/kRXO3pTVyN4eyAv9FX/OdahWs6/xbYMHL/i7ja4GpcrXE6MCRyv987ANp93OHF3gvuB+x3uGkFi7zttwDbxp77Nm4qxHu4KQ07xt9Bvnfu42fTd/Tw//nP043rBrgLPxUnkubT/v2txInqjNjnK/R3mUtkSkyaQXyBRpMiItvhRGuCquYdTTOSR9yk4KWqenStbcka1jxsMkTkHFwtbSGuyXcOrnn4+1raZRihmGg1H4rrr9kK1yy6B/imqr5TU6sMIxBrHhqGkSlsyoNhGJnCmodFEBGrhhpGlVBVCUlnNS3DMDKFiZZhGJnCmoeB2ICFYSSPSFCLsA9NU9MSkb1E5EHvhbJLnBdQwzAyRlOIll9fdgOwA87TwivAlSLyoZoaZhhG2RRsHorIf5aTkar+euDmrCt7B+C/cAtEd8Wtt1urqv3s9avkLwLG4XxWLQB+pKq/iCQ7BOdb6QeqOkNEFuAWo56KW1hsGEZGKNanNZsAHz0eBRITLWB34JRSibyD//vp6xFyT+B/RWQLVZ3u48b440ux4xgMw8gUpZqHUkZIkpdxtafP0es2Ix/T6BWs43FO3R7y11Ojvoo8EjsahpExionWpyLhCJzrjseBL+OaW1/210txfs8TQ1X/qqpTVPVPOJcd/RC3NdYX/eUzqtqhqq/h3KeAc5ebs+t5f8y5390mFm8YRoKoKrfffnsqeRdsHqrq/+XOxe2Jtzlui6DFkfg7cN4CjsL5cqomY3C7nIDrxyLP+V7+OAfnD+lEEXmBXkdq0X4vwzASQFU5++yzufzyy7nttts46KCDEs0/dPQw5wFyeCy+xR8/n4w5ZTE6cv5OgfPNYN2GCEfgalaX45qUp6jqo/ky9htozst3zzCMwkQFa+LEiXzmM59JvIzQyaWrcCNzt4nI/+L6nLamt7N8VaEHU6RQv1Q0ft1Agqo+gvPGWRJVnSkir+J2QjEMI4C4YF122WUVTR4tRaho/RLn9rUVtxFmjpxFtWhmRbe92jByPrJAmrJQ1ZvSeOGG0YhUS7AgXLQm4/xkT8LVuHK8C1yGG8WrNs/jdikeRd+NK6Pnf6s0cxH5XKXPGkYzUU3BgjKdAIrIKNw8qC1xe6Q9rm733GSNEhlKbyf7DbhJpmtxm42C29hzlYj8HJjg447DOf6/AdcMfB8Yo6ovUSFR1zS29tAw+jNQwYqmDXVNU8kuMzsD+6W52wa9O34UCif6dFvgRgXzpZmSgB3r8jMMoy89PT06adIkBXTixIna09NTdh6x71jQ9zJ47aGIHC0iL+G2XLrHx10nIneKyH6h+SSJunlZ+wO/we0avAo3d+wU7Z0NXxHWPDSMwmiVm4RRgpqHIvJp4DZ6O95VVQf7TSbPB36mqmekZmWNsOahYfQnScGqpHkYWtM61x/nxuKv9scDA/PJDFbTMoz+1LKGlSO0ptWNm0i6Fa4DPlfTGoprknWr6qhieWQRq2kZRi9pCFaaNa1cuu5YfG5BsnlANYwGph5qWDlCRWu+P56VixCRPYBZ/vKZJI2qB6x5aBiOehIsCG8engn8iML+tc5S1SuSNKwesOah0eykLViVNA9DRUuAX9HrHSHKb1Q1X3zmMdEymplq1LBSE61IAZ/A+dLaFOdHa46q3l2emdnBRMtoVqrVJEyzpnUC8HtVfbdi6zKG79Na5+XBRMtoFqrZh5WmaPXgPIjeAHQAt6rq2srMzA5W0zKajWp3uqctWtDbEb8MuAboVNWH8j+VfUy0jGaiFqOEaYrW/jh/60fROzcr9+DzQIeqXpjv2SxjomU0C7Wa1pB6R7wvZB/gC8C/A+295engsjLKACZaRjNQ08XPFYhWJTPZH8O5Wt4at6tNQ86Gt8mlRjNQbxNHQwgSHBEZDnwW10Q8DNggdwvnnO+OVKyrIWrulo0GJ4uCBeG1pCX07sST+1SP4kYSr1bV15M2zDCM9MiqYEH5o4cv4hzudajq/CKPNATWp2U0Gp2dnUyePJlFixYBMG7cOG655ZaaCVaao4czgKtU9YGKrcsgJlpGI9HZ2cmpp57KihW9m7YPHz6cWbNmMX78+JrYlIprGhEZBnwJ+JOI7FqxdYZhpE5nZyft7e0MGjSI9vZ2Ojs7192bPHlyH8ECWLFiBVOmTKm2mQMitKb1Gm694UhVXZ66VXWC1bSMLNHZ2cmECRNYvrz3K9rS0sLMmTM59thjGTQofx1FROjp6cl7L23SdAI42x+PKs+k7GJTHoysMWXKlD6CBbB8+XImT57M2WefXfC51tbWgvfqkdCa1oXAV3E7Oc8BngBWRtPYjHjDqC2DBg0q+n86btw47r777j5NxFxNLEt9WqH7//Xg5mMVDKF7lmUpYPseGhmira2t4F6huX0JOzo6tK2tTUVE29ratKOjo6Y2x75jQd/Lcqc8FNE+W8ZjGLUkX58W1H5aQzHSXMazXSUGGYZRPXJNvHqah5UGQR3xqrqwVEjb0IEgIj8UkVdEREWkq9b2GEZaHHvssRx1lBsvmzhxYsMJFoSPHiIim4jIRSLygIg86+PGi8h/ishm6ZmYCINwPu4No2HRDC/NKYuQji9gc+AFXKd7D77jHbjOx/1XaCcasANu67GnfF4KrCmQthW3vnEJznPqY8DJoWXF8hrmy+oq4xnriDcyQU9Pj06aNKlPp3sWoIKO+NAv7ywvMCtiovUpf31vcIFwBP1HN/qJFrAFsDhPWgWmhJYXyc9Ey2hIsipYqpWJVmjz8DCf8Wdi8Q/745jAfABeBi4CPhd5Ph/TcD67AI4HtgRyrp2nisi6GXEi0uX7q+LhT2XYZRiZQ5ulSRghdMrDKtxI4/rAavwUBxHZEHgLWK2qw8ouXGQucCCu5jYkEj8IeAMYBTyjqrv5+GOAa32yb6jqpT5+DDA0TxHvqupin2YYrqa4UFXbA+2zKQ9G3dIIgpXmlIfXcbWevWLxX/HH1wLzCWUMTrAAFkTio+frbFHV54tlJiKfBbb1lyNE5BRggTbwno1GY9MIglUpoc3DOTjnf+uaWyLyCPA9XLNxTsJ2jY6cv1PgvJwRy28BP/XnH8D10X25UGIRmSAi88rI3zCqRjMLFoTXtC4ADsd1jufaSR/CCdlrQNLrDgv9BaLxwe01VR1bTuGqOhOYGW0eGkY90OyCBeGTS18BPgpciROptbgm42xgX38/SZZEzjeMnI8skCZxzMuDUW+YYDmCJ5eq6iuqerKqbq2q66nqVqr6ZVV9KQW7ngfe9uc7ReKj539Lodx1qOpNaeZvNDfFnPXlu9/R0WGClSNkXgRu1HATYAN/vSFwDnApMC50foV/diiuz2o0cB9+nlYkbn2f7uf0zuEYj5vg+qC/Xg1sW0655QbclAybp2UkTkdHh7a0tPSZo9TS0rLO40K++0OGDAmah1VvXhxKEfuOJT65dC0w1V8/RF/XNEcHFwhjo4bmCSdqCpNLKwkmWkYaFHIh09bWVvT+yJEjSwpWMTGsRyoRrdB5Wo8DHwT2AZbjluCsxTkC3AC4X1U/XjIjl9dY4K4iSU5S1dk+bStwMTAOGAHMB36kqr8IKWsg+D6tG3PXIe/JMEIo5Kwv5/a4mDO/Yv+H7e3tLFzY33dBW1sbXV1dFdubJmnuxvMmrkm4Ma7ZdBVuRPE6nIAtU9VNy7a4zrHJpUYalBKXSsWnlBjWI2n6iF/fH9/H1bgUmAc85+NHBOZjGE3P9OnTaWlp6RPX0tLC9OnTAfjud7/LkCFDCt4vRCFf71nzAV+SkDYk8A9cc/AaXD/TWqAd2Aa3YPql0PZoVgLWEW+kSLzD/PTTT1/XlzVy5Mg+x8GDB6/r8yrWP9UsfVqhX+BL6PUT3wP83ccf4a9vCS0wS8FEyxgIoSN5hUYLTzvttLJFyEYPe7+8LbhlMI8DNwE7+/hv4TrV/zO0wCwFEy2jUsqp9RQaLczVsOIhN8rYCFQiWkEd8c2IjR4aA6GczvRyJ4nWc8d6uaQ2eugzF+BY4ACcx4eXgXuBa1S1Md5gDBs9NColdCRPVRk1ahTd3d3BedfzFIZySc01jYhsC9wC7Bq79RXgHBE5VL3fKsMw3IhdvppWdCRP1a0l7O7uZsiQIaxZs6ZkviGjiI1O6JSHGcBuOC8L8bCbv9+UlFpDZjQnpaY15AQrt5bwyiuvpK2trWiebW1tNd0Num4I6fgC3sONHF6F2wNxPX/swI0evhfaiZaVQMCUhywOMRvVo9BIXjGf7r5Lol8QkVp9jFShgo740C9wbm7WqFj8KBp0npb/fEVFq9QaMsOIU2oTimb7n6pEtEKbh1f4456x+D1j95uK3C6+ofFGc6Na2h9WqWalUaQjXkTOj17inP7dLCJ/BF7CzYY/CniV3mU+TUVIZ6thQGnB6uzsZMqUKSxatIhNNtmE4cOH88Ybb9Da2sr06dOtHytKoSoYvTPgQ0LezVazHrA+LSMBSjUJm/n/iAqah6VEKzSsDS0wK4HAtYdZWzZhVJeQjVSbrR8rSiWiVXByqYgUH3/tX2Pr307KODa51BgIqmE+3bPoUiYpKplcWvMaTT0HbO2hEUC+2nY5W9VbTSuFKQ8ub7YFfomb/rAKt4znl0BraB5ZCyZaRiny9UcNHz5cDznkkCDBKpSH9WkNULSAVtwoYc41zbq+LNyWYg0pXCZaRikK1ZJCBStHs/aNViJaoe6WrwRO8Jddvra1NW5WvAK/VtWTSmaUMaxPyyhFMX/uPT09zbvNVyBpuls+GCdOZ6rq9qr6SVUdA5yJm8N1cJm2GkbdUs560mIujk2wUiKkOobrw1oLjIjFj8A1E1eFVu2yEjB3y01Juf1LHR0dOnz48H59Ws3SvBsopNinlVt7eHos/jQvWi+HFpilYKLVfJQ7ktfT07Ou0x3Q1tZWE6wyqES0yu3TUuBFL2Lb4Pq0wPq0jAahnD0HVcPmYRmFSbNP63xgCa7/anvgEzjBEmApMLUcQw2jXinURyUiffq2TLBqSGiVDDftYTbwCrDaH2cDbaF5ZC1gzcOmo6Ojo6BPq1wTMWQtYej0hWad6pCDCpqHNReGagTcztg3eaFdCTwLHB/wnIlWE5JPsMA54kty8XMzTyrNYaJVWHzavVCdA0wE3sINLHywxHMmWhmnkppMoc741tbWRBc/N/PynRyZES1gB2AW8BRu9FEp4N7GN0s7cH1qK4DHgJPLLG8IMChyPcOXeUyJ50y0MkylNZmBLM0px11ys7lWzkeWROuIPH+sfqIFbIEbqcz3x51SYdktvta1Ati2RFoTrQwzkJpMtIbW2toavJbQalrlkSXR2huYDhwOPFREtH4e+VDHeRF70F+vJrLmEbe8KN8/wZ8iaUYAd+CahscG2GmilWGSqMmU461B1fq0yiUzotXHAJibT7Rw0zHe8veejsQfE/mgZ0fixwC75Anb+PujgYeB90MES020Ms9AazLlClYOGz0MJzXRAv6Aa9INDc042IDCorVj5ANdH4n/UCT+qsAyRgDP+Gc6gC/5sHuJ50y0MkhOCHK1qkpqMpUKllEeaYpWzg3NUtzOOx8LLSAg70KitV/kA/06Er99JP7WwDLa8/3iAhcUSD8BmGeilT3yNblywhVakzHBqh5pitZi+vrQWgssAM4FtgstrEDehURr/wKiNSYSP2cgZZewyxZMZ5BaNQmNyqhEtIKW8ajqNr7m8wPgBdzynR2AacA/RORuETlBRIaG5BfIksj5hpHzkQXSJIqq3pRW3kZ6DGQvSlVbmpMFQtceoqoPqeq3VXVH4CPAfTjxEuAAnOvl+SKyR0K2PQ+87c93isRHz/+WUFn9EJHPpZW3kR7F/FsVwwQrQ4RWyVxNjjbc4ugX6et6eQnQ7c/vC8hnKG40bzRO/BRYE4lb36eLTnkYD2xO3ykPRedZDTRgzcPMUck0AmsS1g4qaB6GfnmPBW73whIVq3u9mKyHm0PVDawIyG8sefodIuFEny7xyaXlBBOtbBKfRnD66acXnFZg/rBqS5qileuA7wHewS2D2SNPuvkEbNwaKlo+bSvQiRu5XEkFy3gqCVhHfENQrOYVF6zQmpmRHGmL1mM4T6UjiqTbigZyVWOilX1CFj/nC1leSpOlCauViFao59IDVPW+kgkbCN8Rf2PuOuQ9GfVHMU+kxcjq7s6dnZ1MmDCB5cuXr4traWlh5syZjB8/voaW5acSz6UFRUtEig+3xFDV0mPKGcPcLWef9vZ2Fi5cmPfexIkT+eMf/5h3OkRbWxtdXV0pW5c8hT5vvX6eSkSrVJNwbWDI61Ym6wFrHmaefH1agI4bN057enoabtFy1tzdUEHzsJRohYaSne9ZDCZa1aPcfphyFyW3trb2E6xKy65nsubuJmnRmlpOCC0wKwEbPawalew1WE76ZpqHlbWaY6KiZcFqWtWi3NpBOembSbByZKnmaKJlopVJyu2HCU2fhmBlSRCyQNLNw9BO+IbsiLfmYfVIo6aVlmBlqemVBZIWLeuIN9GqCkn3aaXVJMxaJ3cWSFq0riwnhBaYpWCiVT2SGj1Msw8ra9MJskAlohU0I75Zscml2UI1XfcyWZu4mQUqmVwa7E/LMOqBzs5O2tvbGTRoEO3t7XR2dgLpCxbA9OnTaWlp6RPX0tLC9OnTEy3HKEFolQz4CvB3YDnWEW/UgEJ9WVdddVXVpjXY6GGykFbzUEQmAD/zmef7+VJVHVy5dNYn1jysLwo1z0aOHEl3d7d5HM0gaTYPJ+AEK+feWIHf43ZpfgH4VbCVhlEhhfy8m2A1F6E1rXeB4bituBbia1YichBwC3CCqnamaWgtsJpWfVGspvX222+bYGWQNGtaQ/zxZZzLZURkBHCPz+P8YCsNo0LydYQPGTKEGTNmmGA1EUNKJwFgGc5f+4bA6zgPpT/BNQ8Btk7eNMPoy/jx41FVzjjjDLq7uxk5ciQzZszguOOOq7VpRhUJrWk9449jgLtwnfHH09vXNS950wyjL6pKZ2cn3d3dAGy88cZWw2pCQmtalwN/BQYD5wB7+gCuI/6M5E2rLbbvYX2hqhx22GHMmTNnXdyiRYuYMGECQF26EjbSoaIZ8eJ+3nbEbR32rKquSdqwesA64usDjUwczYfNSM8uifqIj2V8p8tTP53n3oX+3tRwU7OBiVbtKSVYkN1NKIx0Rw/H+pCPc30wjESJCtbEiRMr3vLeaCwGtPZQRD6YlCGGESUuWJdddhkXXXSRrf0zirqmmUq4E8DFoeuGshSwtYc1oZh7GVv711hQwdrDUqK1zsmfD4WcAF4aWmCtAvBn4E1gFfA8cHbAMyZaVaYZfbo3M5WIVrHNWv8NOMJfnuAz/nW0kuZF4K/Ab1W1rntCReR7OLEaDJyHmyC7p6o+UeQZ64ivIqrpu5cx6otEN2vVvjWOLuCFUCUMyG8HYBbwFK6mphRwbwO0Ah3AEtwM/MeAkyssdxNgJ+BRX+buJdJbTatKWA2rOSHJ5mGaAVeDi7ut7SdauKVDi/OkVWBKBeW+5Z/tASYHpDfRqgImWM1LJaIVPLlURA4Gjga2BdbvX2HrP4erSF57e+F6ANdU2we3OcaQWLqf45YKgVs2dDtwPbAv8D6wg6ou8mm7gLY8xd2sqof7NGNx6yTPBzYDPqaq84vYac3DlFG1JmEzk2bz8JsUHjkc0G48wFzy1LRw0zFyNaOnI/HH0KvOZ0fixwC75Anb5CnzW/75b5awrSlqWqVG5NIasatGDctGG+sbKqhphQpLF70jha/56xejIbTAPHkXEq0dIx/o+kj8hyLxVwWWsT9wNXAacCawyD8/rkD6CbhF4A0vWqW240prr79qCZbtU1jfpCla7/pa1UGhGQcbUFi09ot8oF9H4rePxN8aWMZuuM73d4H3gCeB0wKea3jRKrWXXxp7/VWrD8v2Kax/0hStG7xobR6acbABhUVr/wKiNSYSPydpeyLlNMXGFqX28kt6r79qdrrbPoX1TyWiFbqM5xLgHeAPIjJORLYXkdZoCMynHJZEzjeMnI8skCZRVPWmtPKuJ0qt50tyvZ9qdTvdba1igxKibJRexlPxFmIk1BGfdKBJalrV6tOqxbQG69Oqf6igphX6BS60fGfdMp/QAn1+Q4HRPtyXE61I3Po+3c8jH13hZawAABLGSURBVGo8sDnwoL9eDWxbTrnlhmYQLdX0Rw9rOQ/LRg/rm0pEK9Sf1tRSaVR1WsmMevMbi3PbXIiTVHW2iGyBG8XL54P+XFVNbXm/91x6Y+465D0Z/dEqNwmNbFHJPK0gd8vlCFKSqOprIrI/cDEwDhgBzAd+pKq/SLnsm+zLNTBMsIw0KMvdsojsjHMGuImqXpyWUfWCzYivHBMsI4Q0PZciIj/GLXCeAXzHxz0tImtF5IvlmVr/2MYWlWOCZaRJkGiJyBnAV3168QHgUn9+dCrW1RBtkikPSWOCZaRNaE3rNFwP/yWx+Jv9ca/ELDIyiwmWUQ1CRw9X4LYLG4lbCqOqOlhE1gNWAitVtaVYHlnDRg/LwwTLqIQ0+7RW+uOIWPxHY/cbBmsehmOCZVSTUNF61B+vyEWIyNeAa3DNxnkJ22VkBBMso9qENg/H4TaG6HcLJ1qHquptCdtWc2zKQ3FMsIyBklrzUFVvxfmY6qZ39FD89ekNKlg25aEIJlhGrSh3cukGOJcxmwJLgftV9d2UbKs5VtPKjwmWkRSV1LTKEq1mw0SrPyZYRpKktvbQZ57YxhZGNjHBMuqBINESkW8C3y90G9cZbzQwJlhGvRBa0zqTXnF6DefLqqGxjvheTLCMeiJ0ysNbuNnwe6vqo6XSNwrWp2WCZaRLmjPir/bHnjJtMjKMCZZRj4TWtIbhduTZxx8X4dwjr0NVL0zDwFrSzDUtEyyjGqQ25UFEDgV+T/9Rw2iBg0MKzBLNKlomWEa1SFO0ngV2KpJETbQaAxMso5qkOU9rW9zI4deAO2hArw6GCZaRDUJF61bg34Dfq+rrKdpTNzTblAcTLCMrhDYPjwZ+BnQBPyF/R/zdKdhXU5qleWiCZdSKNPu0eig+611VNXhJUFZoBtEywTJqSaprD+ndzMJoEEywjCwSKlonpWqFUXVMsIys0jSuaURkEHAvsB/wkKp+LOCZhmwemmAZ9UKizUMRafUZLcqdF0NVF4UUWEO+DuxZayNqjQmWkXlUNW/ArTNcEzlfWySsKZRPnnx3AGbhdqvOdfDnfR5oBTqAJcAK4DHg5NCyIvnsCCzHeatQ4MHA5zQXGoGenh6dNGmSAjpx4kTt6emptUlGkxP7joV9nwve8EIVOS8W1gYXCEdEDS0kWsAWwOI8aRWYUkZ5uWbhbKC9WUXLBMuoRyoRrWId8RfSO81hWpF05fIycBHwAHAebhF2PqYBW/vz44HbgeuBfYGpInKV+iapiHQBbXnyuBn4A7A7MCmSZn0RaVPVhQP+NBlArUloNBAlO+JFZAiwm798UVW7EytcZC5wIK6mNiQSPwh4AxgFPKOqu/n4Y4BrfbJvqOqlPn4MMDRPEe8CpwBT89xbqKrtJezLfEe8CZZRz6Q1T6sHeAQ3T2s73LZhaTMGJ1gACyLx0fO9cieq+nyhjETkt8CT/nJTYAbwHDAxEUvrGBMsoxEpKVqq2iMiL+KEpBqCBTA6cv5OgfPNQjJS1aeBpwFEpN1Hv6GqtxR6RkQm4PZ5zCwmWEajEuq5NNenNU1E8jXDkqbQtysaX3Z7TVW7VFW0xBwtVZ1Jsv14VcUEy2hkQmfEnwK8jZsyMF5EFtDXPY1qsluILYmcbxg5H1kgTeKo6k1Z/KKbYBmNTqhoHYir2QiwCW4EL0caW4g9jxPJUfR1Phg9/1vCZWYeEyyjGQgVrUUkJEy+eZnrZB8aic/1Y3Wr6ioRuRbXr7SriIzHTXk426d5H7guCXuK2Jkpf1omWEazUPW1hyIyFrirSJKTVHW2iGwBzKN3rlaUc1V1ehr2RcnKlAcTLCOrpLmFWLSQYSKyZbnPlYuqvgbsD/wGWAasAh4HTqmSYGWipmWCZTQbwTUtEdkPuAQ/g11Vh4jIj4ARwKWq+mSx57NIvde0TLCMrJOaE0AR+TBwJ7AefTve1wIn4Gavf7MMW+ueeq9pmWAZzUpo8/BC3J6Hz8biZ+NE7DMJ2lQXqOpNtbahECZYRjMT6iP+Tdx8qe2BF/H7HIrIerj5Wm+p6iapWloD6rF5aIJlNBJpdsQP88dXY/GjY/eNFDHBMoxw0XrRH4/JRYjIKOBSf/lCkkbVA/XWp2WCZRiOUNG6Btd39St6O+GXAl/w19ckb1ptqac+LRMsw+gltE9rPWAOMDbP7f8Dxqnq6mRNqz310KdlgmU0MqlNeVDV1SJyEHAscAjOL9VSnJD9RlXXFHs+i9RD89AEyzD60zRbiFVCLWtaJlhGM5DqMh4R2U5EZonI0yLytj/OFJHtKrDVKIIJlmEUJrRP65PALbipDXFHfCuBQ1T1nlQsrCG1qGmZYBnNRCU1rVDRehT4kL98CbejztbAtj7uUVX9aDnGZoFqi5YJltFspNk83BVXqzpLVdtUdX9VbQPO8vc/WI6hWaDaHfEmWIYRRmhN62lgZ2CUqr4biR+B22ziKVXdIzUra0S1alomWEazkmZN6zxcTeukWPyJwBrgnMB8jBgmWIZRHqE1rbuAPYGNcH1ai4FtcH1aS4GnIsmT3uSiZqRd0zLBMpqdNDvie+jd2II85+uS4j1AhBRe76QpWiZYhpHijHgS3NjCMMEyjIFgM+KLkEZNywTLMHpJrSNeRPYpcX98SD5ZIo0pDyZYhjFwQkcP7xWR6X7PwnWIyGYicj3w6+RNqy1Ju6YxwTKMZCi3I/5J4ARV/buIfBH4CW7HaRql8z1KUs1DEyzDyE+a87R+gNt5Zw/gIRG5B7cf4QeAbuD0sixtIkywDCNZytn38EPAL3FrEHNTHm4GvqKqr6RmYQ0ZaE3LBMswipP2DtM7AVvQK1gKtOMmmRoxTLAMIx1CRw9vBq7GidabwEzc8p3dgPtE5LLULEwIEekSEY2GtMoywTKM9CinIx7gRlxz8HXfXJyNW95T97PgRaQLWI7beBYAVS26IUclzUMTLMMIp5LmIapaMuC2vT8uT/wQ4DvA6pB8Is/tAMzCrVnMjUyuKZC2FegAlgArgMeAk8spz+fTBcwFRpbxjOZCCD09PTpp0iQFdOLEidrT0xP0nGE0K7HvWND3MrSmtZUW6WwXkb1U9W9BKunSHwH8MRa9VlWHxNJtAczDORyMc66qTi+jzC6cAArwHvBjVS3qnSKkppWvFmU1LMMII7UF05ECdsZtI7aJql5cpn3RfPYGjgAewLm92Yf8ovVzYIK/PB64Hbge2Bd4H9hBVRf5tF1AW57iblbVw0XkPGA+TtWn4RwbHqmq1xexs6hoFROlct6rYTQrqYqWiPwYNx8r58lhSMQ54LGqem3ZFrt85wIHEhMtERmEa5aOAp5R1d18/DFArqxvqOqlPn4M0GfGvuddVV0cK/NrwI8oUVsz0TKMdEnNy4OInAF8Nc+tS3EjiUfTKyRJMQYnWAALIvHR871yJ6r6fKGMRGQP4Hu4fRrfB872tx4okH4CvTU8wzDqiNB5WqfhmlWXxOJv9se9SJ7RkfN3CpxvFpjXUn88H7gc16F/iqremS+xqs7UBtyowzAagVB/Wjv641TgG5H4Zf64VWIW9VKoqhjfwqwkqvoq8NmyCq+DHaYNw+hPaE1rpT+OiMV/NHY/SZZEzjeMnI8skCZRNMDLQ6F+K+vPMoz0CBWtR/3xilyE78y+BlfbmZewXQDPA2/7850i8dHz4GkW5RJa08o3j8QwjPQInac1Dvhzvls40TpUVW8LLtT55cp1st8A7I/zIrGFj+tW1VWxKQ/H4aY83EDvlIcxqvpSaLnlUosdpg2jmUhtwbSq3ooTj26cUOVCN3B6OYLlOQDXtFuCEyyAwZG4//BxU3G7WYObFf8aTrAApqUsWNanZRh1SGhHPKr6CxG5Bicym+JG5O7XyOatSaOqr4nI/sDFwDhcn9p84Eeq+ou0yvVl32Qz2g2j/rCNLYpgzUPDSJc0txBrOuLNQ6t1GUZ9YDWtIqTpc8swjL4k2hFvGIZRL1hNqwkQkXm2LKky7N1VTlrvzmpahmFkChOt5mBmrQ3IMPbuKieVd2fNQ8MwMoXVtAzDyBQmWoZhZAoTLSMvIjJIRO73e0Q+WGt7soCIbCwiN4nIKyKyUkSeFZHja21XPSMie4nIgyKyyu9NemKpZ0y0jEJ8HbenpRHOKJzDzB8D/4XzWjJbRD5YU6vqFBFZD+e1ZQfc/9srwJV+T9WCmGjVOSKyg4jMEpGnRKTH13zWFEjbKiIdIrJERFaIyGMicnIFZe6I28/yvwdqfy2pwbtbDOymqher6g+B3+C+Yw0hWim8z0OAbYErVXUGzh06wKlFDQndINFCbQJuqzWNhX4b2+J+1RfnSavAlDLKGwTci9s9vN0//2Ct30MW3l0szxbgWdx+BNvW+l3U4/vE1a4UOMtf7+yv5xSzw2pa9c/LwEXA54CHi6SbRu+mtscDWwIP+eupItKaS+j7DjRP+BNwIrA78BN695FcX0Ty7SlZ71T73eXSjABuwjUVT9YU/b5VmcTfp0dix+LUWr0tlPVLN5c8v2642tFb/t7Tkfhj6P2FOzsSPwbYJU/YBriA/L+QXbX+/PX+7vz90bgv9Pu4/UBr/tnr9X0Cn/fXP/DXn/HXVxQr21zTNAZJ7hH5W+BJf7kpMAN4DpiYiKX1R5LvbgRwD07EOoEeEfkS8KSqPlnouQajnPc5B9eMPFFEXsC5VAco6uDTRKsxSGyPSFV9GngaQETaffQbqnrLAOyrZ5LcX3M0TrAAxvsArrnULKIV/D5VdbWIHIH7YbwceBW3H+mjFMFEqzFIbI/IKKraVSTvRiHJ/TW7iuTXLJT1PlX1EXr3fQjCOuIbg5ruEZlx7N0lS+rv00SrMajpHpEZx95dsqT+Pk206hwRGSoio0VkNDA0Ej/ah/VVtQe41t/aVUTGi8jmwNk+7n3guupaXnvs3SVL3bzPWg+dWig5tDyW/FMQcuFEny7xCZJZD/buGvN9Wk2rQVDV13B7Uv4GWAasAh7HjcZMr6Vt9Y69u2RJ+32aE0DDMDKF1bQMw8gUJlqGYWQKEy3DMDKFiZZhGJnCRMswjExhomUYRqYw0TIMI1OYaBmZJOJBdG4kbq6P6yozr7EicoEPGyVsZ0U2GYUx1zSG4ZanTPXns3GeN406xUTLqCp+26g16hbWJoqqjk06T6P+sOahUTbRJo+IHCgij0Q2Jz06ki636cNsEfmGiCwCVuL9LInIviJyo4gsE5HVIrJARM4TkaGx8g4WkSd8GfNE5IBSdsXiPywivxOR1305i0Xkmtwz9NayAF6M5yEiu4jI1SLymn9+oYhcKiIjY+V8RNzGoytF5GnvldNImlqvHLeQvUDvpgbv4bbIiq7iXwt83KfLxb0RS7MRMA5YTX5PADdFytoVt+A2ev9doNufz81jV1ck7qBC5cSeybuRB/AvkbLiYR6wvk/3AeDN2P01OId3md8YpJ6C1bSMgdACXIEToaNwX85BOJ/oUTYGJuM2PNgVJ3ZX4Hwy3Y/bX3E4bh88gMNF5BB/PgVYz59P8Hn8EBgRaONPfTk9wGne1jbgPFjXpIzau52qiqq2++tLfVnP4TZdHQb8u7/3EeAkf/51nzfAhd7Or9HXZ7qRBLVWTQvZC/TWTlYBwyPxd/v4FTif4Lkax5Ox53cif80lGr7v077gr5+PPD+M3trX3Dx2deUp53dFPs8FkXTtkfgWXG2pmJ3X+rR3+uvlwHqRPBZiNa1Eg9W0jIGwTFVXRK5f9sdhwCaR+PhONJsG5J17fstY3qjqSlyTsxTRcuYHpI+zMTC4RJq4nUtVdXXk/isVlGsUwUTLGAgfEJFhkevcrsJxUVkZe25p5Pxydc2xPgHXFAS3rVQ0b3yZUVEsRHQDhZ2LpCvkVO5NXLMS4PoCdh4cs3O0HyHNsVWAnUYZmGgZA2E94EIR2VBEjgQ+7uPvV982KsAC4EV/foqIHCoiw0RkUxH5kog8jOt3ArjPH7cXkVNFZENcf9R6lEBVFwD/8JdH5p4XkW1E5JxI0jcj57tHnl+Oa/ICfE5EjheRDURkIxE5XET+AnwyZudwYIov53QgvgW8MVBq3T61kL1Ab99RN24kr9To4ew8eRyG2+SgUF9Ru0+3C/1HD1fgOvMHPHro0+yX536Hv/fhPJ8xGsb6dPlGD3sicV1JvHsL1qdlDIxluObRX3HCsgD4oqreW+pBVf0z8AngRp/PauAl4BZc0/AVn+5Z4PPAUz7N33GCF7R3nqr+BfgY8Af/zPu4/rHfRtI8gBvdfIne5mDu3qPA3sDVwOv++VeBu4BJwKM+3TKcQD7s7VwAfAl4LMROIxzzEW+UjZ+QeSCwUHunBhhGVbCalmEYmcJEyzCMTGHNQ8MwMoXVtAzDyBQmWoZhZAoTLcMwMoWJlmEYmcJEyzCMTGGiZRhGpvj/1wDVUCi/v5EAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAS0AAAEmCAYAAAA3JMYtAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO2deZwcVbn3v7+QBBgTgiQsIswMhl3kFWUR3HjvBYMKyvUieAkovGDcDaLXLWqA6+i9XkVcQIGrRpOwuF0QkEXEqCggAWWLgAYmIWyGPSGQbZ73j3OaqXS6e6p7qrq7up/v53M+VXXq1DlPV0//5pynTj1HZobjOE5RGNNqAxzHcerBRctxnELhouU4TqFw0XIcp1C4aDmOUyhctBzHKRQuWjkj6TRJJulvVc7/PZ4/rc5696/3msS1B8c292rk+rK6Toh1TRhtXc7oid/Fh1ttR564aDWH54GdJO2bzJS0H9AXz9fL/sDsDGxzOosDgZ+02og8cdFqDs8C1wHvKst/V8x/tukWOR2FpM0BzOxGM3u01fbkiYtW87gIOFqSAOL26Ji/EZJeJ+m3klZJelzS+ZImxnMnAN+K+xbTgni8u6SLJD0Qr71L0imSKn3X20u6XNKzkpZKen8FO46WdIek1bHOAUlja31QSVMk/TDavUrSggq9zEFJX5X0aUkPS3pa0tcUeEu0e4WkSyS9uOzarSSdK+lRSc9L+qOkA2rZFK+bHK97OF53j6RTEud7JH1T0iPx/M2S3lRWxwJJP5V0oqT7Ja2UNFfSpnHI/qeYt0BSb+K6/vg9HRvLr5D0D0mzy+of8ftLDO+nSfqFpJXAt+O5DYaH8e/o95Keiekvkt6ZOL+JggtjafyO75J0bJlNcyQtlHSopNvj38v1kl4+0j3PBTPzlGMCTgMeA7YEVgOvj/lvIAwLt4znT0tc89pY9mLgLcDxwIPAT+P5rYGvAga8JqY947l/Bk4HjgAOBk4BngY+k6j/4HjtA8CXgGnAuTHv8ES5N8W8HwKHAZ+Mdn03UeaEWGZCIu964BHgxGjH74AVwM6JMoPAMuDnse5ZsZ6vA7cA7wCmA0+WtbcpcCtwH/DueO2lsf7tanwPmwN3AI8CHwT+CZgBfCVRZn6s5yPAm6Nta4HXJcosiHYvAA6Pda0GzgNuizYfCSwFrkpc1x8/34PxXk8DBoAh4EOJcvV8f8uA/4if5aB4zoAPx/0tgKfi93do/D4/Drw3UddA/IyfizadF+v4t0SZOcA/gL8AxwBvA+4F7gLU9N9Uq3/UnZ6IohX3LwXOjvvnAJfE/XLR+j3wm7J6/in+Me0Vjz8M2AhtCxgLfBa4L5Ff+qM/r6z8r4AbE8c3VrDjk8B6YId4fAIJ0SKIiAFvTFzzImA5cG4ibxD4O7BJIu9PwDpgp0TeV4BHE8cnAWuAXRJ5Y4HFwH/XuBfviwLxyirn94jn35PIGwPcCVydyFsQhWBSIu/H8TO/IZH3wZjXE4/74/E1Ze2eTxCyMQ18f1+vcE1StPaNxxOrfOatCK6J2WX5vwTuSRzPid9L8p4fGevevdm/KR8eNpeLgKMkbQocRYWhoaQegjP1x5LGlhKh97IWeHWtBiRtJul0SX8n9ADWEv6b7lRhWPe/Zcc/B14dhwybAK9iY6fuxYQf84FVTNgfWG5mvy1lmNmzwOXA68rKLjCz9YnjvwODZnZ/Wd7WksbH40MIPbH7E/cG4LeEH2k1/gn4s5n9pcr5/Qgi8cLnNbOheFxu90Ize7rMxjWE7yiZB7B92bWV7vn2wA5Q9/d3RZXPUmIxsBK4QNLbJW1Zdn4voIfK3/GukrZJ5A2aWfIJ+KK43WEEGzLHRau5/AKYQPgjfBFwWYUyLwY2IfTE1ibSamAcsOMIbfwX8AlCN/8thB/jF+O5zcrK/qPC8VhgSkzjCMOpJKXjraq0/5IK15SuK7/mqbLjNVXyBJREawphOLy2LJ1I7XszGXi4xvmXACvNbFUFu3viP5padq+IIpfMg3T3vNQ+1Pf91XS4m9mThCHhOEJvcLmkKyS9rKzNat9x0pdY6TNXsil3ajpUnWwxs2clXQ58DPhJ7IGU8xSh230aoZtezkMjNPNO4Ftm9pVShqS3Vim7TYXjdYThKgQxKC+zbdw+UaXOhytcU7qu2jX18ASwEPhAhXOra1z3OLBzjfMPAxMk9ZQJ17bAKjOrVXc9VLrnpfahvu9vxLhSZnYDcJjC08VDgDOBCwjCX2pzG8L9KTHSd9xSvKfVfL5D6GF9t9LJKGQ3AruZ2cIKqSRaayAMJ8qq2JzEjzcO88qnWpT4lwrHt5jZ+jhsu4XwI0pyNMH3c0OVOm8CtpH0hoQNPcBb2XD41Ci/JojP0gr35o4RrttH0t5Vzt9MEIGjEnYrHmdhd4nye/4Ogngsi8f1fH+pMbPnzOwy4PvAnjH7TmAVlb/je81s+WjbzQPvaTUZM1tAcObW4pPAryUNAT8lPNHqJfzwZ5nZvcDdsexMSdcBz5jZPQRn+oeiT+QJ4EOEJ26VeLOkAYI/6B2EJ0xvT5yfDVwt6QcE/9srCE+rzjezZeWVxc93taQ/ABdL+jThP/gnCD/G/x7hc6fhR8D7gQWSvkp4ijiZ4Et7xMy+XuO6DwHXKLxJcA+wE7CrmX3azP4q6ULg25K2IPik3gvsTuVeXaO8XNK5wM8IT5BPAmYmhpb1fH81iT20/wdcQnia+VLCA4nrAMzsCUlnAZ+TtI7Qg30HYVj6b419vCbQbM9/tyUSTw9rlNng6WHMOwC4CniG8IRnEaFrPymeF+HJ2kOEns+CmL8twdn7DME38RXCjy/5hO/geDwNuJLw33YZ8MEKth1DmCqwJpYZAMYmzp+QrDvmbU0QiSeB5wiiuF9ZvYPAV8vy5hCc3IxQ/yTgG4QpGyW7fg68doT7PJnwtO4fhOkmdwMfTZzvIcx/e5TQ21kITCurYwFx6kmt7zhxj0tPe/vj8XTgQsI/ouWE6Q1KXFfP97dXhc+YfHq4G+Gf3gPx8ywj9PC3SpTfJNpQupeLgOkpvpfS5zm83Ia8k6IBjuPkiKR+4H7gCDO7vLXWFBv3aTmOUyhctBzHKRQ+PHQcp1B4T8txnELhUx5qIMm7oY7TJMxMacp5T8txnELhouU4TqHw4WFK/IGF42RPeFOqPrqmpyVpH0k3xuiMgwrRPx3HKRhdIVoxFtOlhBdtP0Z49eUHkl7ZUsMcx6mbqsNDSe+upyIz+9HozXmh7Z2BTwEHESJKClhvZhvZG+Nwl0IGTyCEgf2mmX0vUewwQqylr5rZOZLuJbyY+l7CC6mO4xSEWj6tOaSI1xMxwguyWbEXcPJIhSRtB/yR8PZ6ib2B/5G0nZkNxLypcftA2XYqjuMUipGGh6ojZcmDhN7TEYS44dU4nWHBOp4QifGmeDw7uRpKRGVbx3EKRi3R+r+JdCRhdZXbCfF5Dovb2wlhVY6qUkdDmNnNZjYrvg3/XKUycUmlY+LhX81snpk9QgjfAiHEbMmuxXFbCse7Q1m+4zgZYmZce+21udRddXhoiYUJJM0hxPnZ3xLB3yT9GlhCCBxWHrA/b6YS4ipB8GNRYX+fuL2KEEvoBEn3AcfF/KTfy3GcDDAzTj31VM466yyuueYaDj300EzrT/v0sBQidvOy/J64fVs25tTFlMT+M1X2twEwszWE3uJi4CzCkPJkM7u1UsWSZkhamK25jtP5JAVr5syZHHLIIZm3kXZy6WrCk7lrJP0Pwef0Uoad5VkF/a+Han6pZP4LDxLM7BZCNNARMbPzJD1MWD3HcZwUlAvW17/+9YYmj45EWtH6PiFueS9wRiK/ZFErhlnJoPtbJPYnVilTF2Z2WR433HE6kWYJFqQXrc8S4kefQuhxlVhJWMb89IztSsNiwnLhk4BdE/nJ/T83WrmkIxq91nG6iWYKFtQZBFDSJMI8qJcQlj263TZcaTcbo6RxDDvZLyVMMl0PbBfzVpjZ6riqyYyYdxxwbSx/AGHNvqlm9gANkgxN4+8eOs7GjFawkmXThqZpZHWZ3YAD81xtg+HVRqqlE2K57QhPBSuVmZWBHS/U5zjOhgwNDdkpp5xigM2cOdOGhobqrqPsN5bqd5n63UNJR0l6gLDE0O9j3k8kXSfpwLT1ZImFeVkHEVbMfZzwQOB2wpPBgVrXjoQPDx2nOtbkIWGSVMNDSf8MXMOw493MbJO46OUXgO+a2Qdzs7JF+PDQcTYmS8FqZHiYtqf1ubhdUJZ/Ydy+MWU9hcF7Wo6zMa3sYZVI29NaQZhIuj3BAV/qaY0jDMlWmNmkWnUUEe9pOc4weQhWnj2tUrkVZfmlF5I9AqrjdDDt0MMqkVa07onbj5YyJL0COD8e/jVLo9oBHx46TqCdBAvSDw8/DHyT6vG1PmpmZ2dpWDvgw0On28lbsBoZHqYVLQE/ZDg6QpILzKxSfuFx0XK6mWb0sHITrUQDryfE0tqaEEfrKjP7XX1mFgcXLadbadaQMM+e1nuAn5nZyoatKxjRp/VClAcXLaeTmT9/PrNmzWLp0qXsuOOO7Lnnnlx11VW5+7DyFK0hQgTRS4F5wNVmtr4xM4uD97ScbmD+/PnMmDGDVatWbZA/bdo0rrzyynxffs5ZtGDYEf84cBEw38xuqnxV8XHRcrqB/v5+lixZslF+b29vxfwsyVO0DiLEW38Hw3OzShcuBuaZ2RmVri0yLlpONzBmzJiKf9+SGBoaqnBFduTuiI+N7A+8E/hXoH+4PdukrooKgIuW0w309fWxdOnSivmDg4O5tp3njPgktxHWGrwRWNfA9YXAJ5c63YCZseeee26UP378eFauXMmYMWPo7+9n/vz5LbCuMqlES9LmMTTNRYQQxj8lLN81lhCc71f5mdgazOyyVtvgOHlSmtZw1VVXMW3aNHp7e5HE5MmTMTMef/xxzIwlS5YwY8aMthGutD6tlQyvxFPqwt1KeJJ4oZk9mo95rcWHh06nUmseVjXHfB7DxWY8PbyfEHBvnpndU+OSjsBFy+lERpo42kzHfCOilTY6w3eBuWZ2QwN2OY7TJqSZ6V5tqkNvb+9Gea1gRJ+WpM2AdwGXS9ojf5Mcx8mDtK/mDAwM0NPTs0FeT08PAwOjimCeHWkCyQOPEBzuPWmDz3dCwhe2cDqEehehmDdvnvX19Zkk6+vrs3nz5uViV9lvLNXvMq1P6z+BfwfeY2bzclHPNsPfPXQ6BWuzeFhJ8vRprQGeAn4g6RjgDuD5ZAHrsBnx5itMOx1AOwtWo9Tz9LBmQfMZ8Y7TVhRBsPLsacHw/KxK+C/acdqIIghWo6QVrZ1ytcJxnMzoZMGCBl6YLiKSvkF4yfslwBIz6095nQ8PnUJRNMHK9YVpSVtJ+pKkGyTdHfOmS3q3pG3qtra5jCHEuHecjqVogtUwaeZFANsC9xHmag0B62P+T2Lep9LOsQB2Jiw9dlesy4B1Vcr2Et5vXE6InHobcFLatsrq2iy2NVjHNT5PyykE9c7DahdoYJ5W2p7WFwmxs9aU5Z9DcNDXE8ZlL+BkYE9qOPclbUcIgTMdmEIQnb2B/5E0q472HKejsbIe1r777stOO+3UlmFlMiGNsgEPEnpUr2XDntaL4vHDaVUS2A8YAA4HbqJKTws4l2EVPg7YjhDDywji2ZsoO5gom0yXJ8p4T8vpOMp7WHPnzrWenp4Nfgc9PT25zWgfLTTQ00r7410dRWtsmWhtEY+fT9tgWb0LKokWwQf1VDy3KJF/dOJDnprInwrsXiHtkCjjouV0FJWGhH19fZX+eVtfX1+rza1II6KVdnhYipe1T1n+++L2kZT1pGUqMCnu35vIT+6/YIuZLTazuyukZQCS3gqcEItPkHSypDdkbLPTZObPn09/f3/nDoNqYFbZ6V4pbDJQNb+QpFE24DxCj+rRuF0P3BK364HvplXJsnoXULmndSDDCvyjRP7LEvlXN9BOMs2pUX4GsBDvabUt8+bNK9QwKEtqOd27oaeV9ke/PfBQQrBKaSjmb5+2wbJ6q4nWQVVEa2oi/6pG2qzTPhetNqVoP86sGOkpYdHEPDfRsmHh+h7BKb8mitX3gR3T1lGhzmqitUviw1ySyH9lIn9uo+2mtO0IF632JU783ShJarVpuZF2WkOzwspkQa6ilUeqIVp1OeJztM9Fq03ptp5WUedhjUQjopV2NZ5N44z4F8XjLSR9RtKZkqalqSNR1zhJUyRNAcYl8qfEtKmZDQEXx1N7xJn32wKnxry1hImtueFLiLU3bR9dM0PMumSme1rSKBthBvt6YHY8vokNfVtHpVVJ4GAqz6kqpRNiue2AZVXKzErb3mgS3tNqa9ptGJSHPZ3awypBAz2ttD/e26M4vRrYg+CAXwusiPvXp24wpWjFsr3AfOAxQtDBhl/jqTfhPi2nDvJwgHe6YJnlK1pPRtHagvBazRDwhYSALU/bYJGSi5aTlqx9bN0gWGaNiVbayaWbxu1a4OWxkYXA32L+hJT1OE5HkuWkTjP3YdUirWg9FLc/AN4d9xcR/E4Qhm8dhTvinXqotiZgvWsFVhKsCy64oGtn/lckTXcM+BrDE0uHgL/E/CPj8ZVpu3ZFSvjwsCtpxKGehU+r0pCwaJNF64UcfVo9wHcIDvnLgN1i/r8DvwHenbbBIiUXre5jNCIxmqeH1XxYnT4frRHR6opwy43g6x52J/39/RWXhO/r62NwcDCXNs2q+7DGjBlT8W9PEkNDQ7nY00waCbecWrQUaj+WEFPrpYTXea4HLrIwGbTj8Bjx3UezRaKWYEFrRLSZNCJaaYdJOwJ3suGE0lK6g0Tcqk5K+PCw62jmcCzNtAb3aVX4XaYqFPxYQ1XSeuAXaRssUnLR6j6aJRL1zMNqt5n/WZKnaD0bxWkuYQ3E8XE7LwrXs2kbLErCZ8R3LXmLRLdMHE1DI6KVyqclaRlhzcCtzOzpRP4kwmz5B81sxxErKhju03KyxswnjibJc93Ds+N277L8vcvOO45TBResbBhb7YSkLyQPCaGWr5D0v8ADwA7AO4CHGX7Nx3GcCrhgZUfV4aGk0kKqaTAzqyqARcWHh04WuGBVp5Hh4UhC07V31t89dLLABSt7avW0+uqpyMw2ngFXcLyn5YwGF6yRyXVGfDfiouU0igtWOvJ8eoikHSV9X9IySaslPRiP64u94TgdjgtWvqSdp9VLiAu/DRv6uQxYDuxvZh20hG3Ae1pOvbhg1UeePa3TgW0JgjVIeFH6/ni8dTzvOF2NC1ZzSCtabyL0qj5sZi8zszeY2VTgwwThelNeBjpOEXDBah5ph4erCdMjJpnZykT+BOAZYK2ZddQEU4+n5aTFBatx8hweLo/b48vyj4vbjosRb2aXtdoGp/1xwWo+aUXrV4Rh4Lcl/V3SAkl/J7xzaMA1eRnoOO2KC1ZrSDs83JGwZNjWMat0kQi9sP386aHTTbhgZUNuw0MzewDYD/gR8Aghttaj8bgjpzs4TjVcsFpLV8yIl/RigsC+GtiKMG1jwMzmjnCd97ScDXDBypZcZ8QXnEnALsC3gE8RFpmdI+nlLbXKKRQuWO1BS0RL0s6Szpd0l6QhSSZpXZWyvZLmSVou6TlJt0k6qc4mlwF7mtmXzewbwAWEz+6iVQDmz5/f8hWWXbDaiLRxmbNMhJWpy1c8WVeh3HYEwam0QsqsBtvuAe4GngN2HKGsx4hvMe2wGo3HdM+Pst9Yut9w2oJZJoJTfwA4nPBOYzXROjfxoY6LInZjPF4D9CbKDlYRt8sTZSYAvyY8SDg2hZ0uWi2m1Sssu2DlS2FEawMDYEEl0SIM356K5xYl8o9OfNBTE/lTgd0rpB3i+SnAn4C1aQTLXLTagvgwZKMkKfe2XbDypxHRSuXTkvRzSUdKGpemfEZMJTjQAe5N5Cf39yntmNliM7u7QloWXzf6PaGHdzEwJOldkvbK+0O0imb6gfJsq7e3cuSjavlZYeY+rLYljbIxvCjrY4RZ8K9Jq4op6l5A5Z7WgQyr8I8S+S9L5F+dso1+Kvy3Bk6rUn4GYTJtIXtazfQD5d1WK3xa3sNqHjTQ00orLMvYcEXp9YQez+eAndI2VqXuaqJ1UBXRmprIv2o0bY9gV2EXa22mH6gZbTVzhWUXrOaSm2iFujkA+ArwtwoC9jvgPcC4tPUl6q0mWrskPtAlifxXJvLn1ttenbYVUrSa6Qdqpc8pa1ywmk8jopV6npaZ3WRmnzSzXQgzy/9AePdQwGuB7wP3SHpF2jpHYDFQWs1610R+cv/PGbW1EUVejaeZfqBW+Zyyxsx9WIUhrboFUaQPmE2IWrqe4R7XcmBF3P9DinrGEZ7mTSGInwHrEnmbxnLJKQ/TCdFTk1Meas6zGm2ioD2tTvJpNQPvYbUOGuhppf3xHgtcG4UlKVbXRzEZT5hDtQJ4LkV9B1NhSJFIJ8RymU8urScVVbTMmusHamZbWeOC1VryFK2S/2qIEKn0HOAVFcrdA6xPUV8q0Yple4H5hCeXzwO3ASel/YCNJgrsiHfS4YLVehoRrbTxtIaAO4DvAPMsEXK5rNz2BGd8Ryzc6lEeOhcz92G1A41EeRibsu7Xm9kfRipkZg+lrK/tKbIj3qmNC1axqdrTqncRVuvAQIDe0+o8XLDai0Z6WrVEa4gw1kyDmVnaXlthcNHqLFyw2o88ggCqjuQ4TaWedx5dsDqHWr2jrl412n1a7c38+fOZMWMGq1atAmDJkiXMmDEDgOnTp29Q1gWrs+iKGPGN4sPD9qW/v58lSzZ+SN3X18fg4OALxy5Y7U2mPi3HRaudGTNmTMXvRBJDQ0OAC1YRyNSnJWl9HalifPci48PD9makdx5dsDqXWo74epzwHffXYGaXtdoGpzoDAwP09PRskNfT08PAwIALVodTyxH/w6ZZ4Th1UnK2z5o1i6VLl9Lb28vAwADHHnusC1aH4z6tGrhPq1h4D6t4+GKtTtfigtU9pBYtSe+T9BdJq9wR77QTLljdRdooDzOA7xJe66n012BmtknGtrUcHx62Py5YxSbP4eEMgmCVwhsb8DPCKs334U57J0eqva7TasFq5jJtToI0QbeAlYQggDsSAwLG/EMJ0Uynpw3gVaSEBwFsOdXCOc+dO7elAfw6Icx0O1D2G8s0CODzhLju4wjRQzchLKS6DlgF/M3MdstAQ9sKHx62nmqv60ycOJEVK1a0bEiY9jUipza5vcYj6UFCvPbJhAim2wNzCcPD9wGrzGxC/Sa3Ny5arafa6zpAS31YaV4jckYmT5/WX+N2KvAbgjP+eIZ9XQtTW+k4dVDtdZ2JEye21OneKUunFZG0onUWYaHWTYDPALcz/PrO/cAHc7GuhRRxykMnOoYrva4zduxYzjnnnJY+Jaz1GpGTM2mdX7ahg1qERVP3AsY2UkcREgVyxHeyY3ju3Lk2ceJEA2zixIk2d+7cVptkZsVeOq1dIEdH/HWx0n+ucO6MeG52BhraVhTJp9WpjmEzn4fVyeTpiB+iygTSWueKTpFEqxMdwy5YnU/T3z2U9PLRXO9kR6c5hl2wnGrUCgI4u/RuIXFVnvJ3DgkOeQMebo65TjU6yTHsguXUpJqzC5hNmP0+RJgNvz5xXJ7OTOtEa1UCfgk8CawGFgOnprimMI54s85wDPtS9d0FWTriJb0dODIevidW/KOk3kURuBn4sZm1teNE0n8SxGoT4POECbJ7m9kdNa4pjE+rEzDvYXUdjfi00vZSBoH70iphivp2Bs4H7iL01AxYV6VsLzAPWE6YgX8bcFKD7W5FmKpxa2xzrxHKF6qnVWS8h9Wd0EBPq1VDtSOTxlYTLcKrQ8sqlDVgVgPtPhWvHQI+m6K8i1YTcMHqXhoRrdThliW9CTiKEOlh0407bBvP4apR135RuG4gDNX2J0SOGFtW7lzCq0IQXhu6FrgEOABYC+xsZktj2UGgr0JzV5jZ4bHMwcBLgS8A2wCvMbN7atjpw8OcMfMhYTeT5/DwEww748vTC6FqGknAAir0tAhPNks9o0WJ/KMZVudTE/lTgd0rpB0qtPnv8fpPjGCb97RyxHtYDg30tGqtxpPkwwxHLP0HITxN3l2PqYTwNwD3JvKT+/uUdsxscbWKJB0EfAT4LWEFoo/EUxWd8DFS64xK55xsMO9hOQ2SVrSmEETqMDP7VY72lLdZ4pkq+9ukrOspYDfgCIZf8v6AmV1dqbCZnQeclxweOtnhguWMhrSi9WvgcMJk0mZR7a84mZ9KVMxsEfCquhovYJSHIuCC5YyWtK/xfI3Qw/m5pGmSXiapN5lysG15Yn+LxP7EKmUyxXyF6cxxwXKyIG1P6zdx+xrCzPJyrI660rIYeJrg19o1kZ/c/zM54T2tbHHBcrIibU9LKVJqJI2TNEXSFELc+VL+lJg2tTDD/uJ4ag9J0yVtC5wa89YCP6mn3XrwnlZ2uGA5WZK2d3R6xu2+luHeW4lNGB7unQjMIbz/+FbC3Kp55TaZ2QMZ2/UC3tPKBhcsJ2tSiZaZZS1aqTCzR+J0hS8D04AJwD3AN83sezm3fZn/uEaHC5aTB6lnxANI2g04GNjKzL6cl1Htgs+IbxwXLCcNuQYBlPQtwgvO5wD/EfMWxdhax9Rnavvjw8PGccFy8iSVaEn6IPChWD7peD8z7h+Vi3UtxB3xjeGC5eRN2p7W+wnTGr5Wln9F3O6D0/W4YDnNIO3CFs8B4wkTO1cSF7KQNJ7wHuLzZtZTq46iEYeHvygdu0+rNi5YTiPk6dN6Pm4nlOXvW3a+Y/DhYXpcsJxmkla0bo3bs0sZkj4CXEQYNi7M2C6nILhgOc0m7fBwGpVf3xFBtN5sZtdkbFvL8SkPtXHBckZLbsPDGMJlBrCCDV/dWUEI8dKJguVTHhLMnz+f/v5+xowZQ39/P/PmzXPBclpCvZNLXwQcBGwNPAb80cxW5mRby/GeVmD+/PnMmDGDVatWvZA3duxY1q1b54LljIpGelp1iVa34aIV6O/vZ8mSJRvlT5w4kaefftoFy2mYXEUry4UtioKLVmDMmDFVP3833xdn9DQiWqlemJb0CeC/qp0m/3jxTgvp7aZS6SQAABD6SURBVO2t2NPq66u0+JHj5EvaKQ/JhS0eAZYm0pK47SjcET/MF7/4RcaO3fD/W09PDwMDAy2yyOlm0sbT2pLQm9rPzG4dqXAn4KFpAmbGLbfcwrp165g4cSIrVqygr6+PgYEBpk+f3mrznC4kbU/rwrgdyssQp/0on4f19NNPY2YMDg66YDktI+3k0s2ASwkrQV9KGA6uS5YxszPyMLCVdLMj3ieOOs0gt6eHkt4M/IyNnxomG9wkTYNFoltFywXLaRZ5itbdbLgKTjnmotUZuGA5zSS3KQ+EuVlGWE7+13RgVAfHBcspBmlF62rg7cDPzOzRHO1pG7ptyoMLllMU0g4PjwK+CwwC36ayI/53OdjXUrpleOiC5bSKPH1aQ9Se9W5mlvUK0y2nG0TLBctpJXn6tKDOVaSd9scFyykiaUXrxFytcJqOC5ZTVLomNI2kMcD1wIHATWb2mhTXdOTw0AXLaRcyHR5K6o0VLS3t18LM2v2l6Y8Be7faiFbjguUUHjOrmAjvGa5L7K+vkdZVq6dCvTsD5xNWqy45+CteD/QC84DlwHPAbcBJadtK1LMLsIoQrcKAG1NeZ6XUCQwNDdkpp5xigM2cOdOGhoZabZLT5ZT9xtL9nqueiEKV2K+V1qduEI5MGlpNtIDtgGUVyhowq472SsPCOUB/t4qWC5bTjjQiWrUc8WcwPM3h9Brl6uVB4EvADcDnCS9hV+J04KVx/3jgWuAS4ABgtqS5FoekkgaBShHprgB+DuwFnJIos6mkPjPbOLJdB2I+JHQ6iBEd8ZLGAnvGw/vNbEVmjUsLgDcSempjE/ljgCeAScBfzWzPmH80cHEs9nEzOzPmTwXGVWhiJXAyMLvCuSVm1j+CfYV3xLtgOe1MXvO0hoBbCPO0diIsG5Y3UwmCBXBvIj+5v09px8wWV6tI0o+BO+Ph1sA5wN+AmZlY2sa4YDmdyIiiZWZDku4nCEkzBAtgSmL/mSr726SpyMwWAYsAJPXH7CfM7Mpq10iaQVjnsbC4YDmdStrIpSWf1umSKg3DsqbaryuZX/d4zcwGzUw2whwtMzuPbP14TcUFy+lk0s6IPxl4mjBlYLqke9kwPI1ZtkuILU/sb5HYn1ilTOZYQWPEu2A5nU5a0XojoWcjYCvCE7wSeSwhtpggkpPYMPhgcv/PGbdZeFywnG4grWgtJSNhisPLkpN9XCK/5MdaYWarJV1M8CvtIWk6YcrDqbHMWuAnWdhTw85CxdNywXK6haa/eyjpYOA3NYqcaGZzJG0HLGR4rlaSz5lZ7ovuFWXKgwuWU1QamfKQ1hGfbGQzSS+p97p6MbNHgIOAC4DHgdXA7cDJTRKsQvS0XLCcbiN1T0vSgcDXiDPYzWyspG8CE4AzzezOWtcXkXbvablgOUUntyCAkl4FXAeMZ0PH+3rgPYTZ65+ow9a2p917Wi5YTreSdnh4BmHNw7vL8ucQROyQDG1qC8zsslbbUA0XLKebSRsj/knCfKmXAfcT1zmUNJ4wX+spM9sqV0tbQDsOD12wnE4iT0f8ZnH7cFn+lLLzTo64YDlOetG6P26PLmVImgScGQ/vy9KodqDdfFouWI4TSCtaFxF8Vz9k2An/GPDOeHxR9qa1lnbyablgOc4waX1a44GrgIMrnP4tMM3M1mRrWutpB5+WC5bTyeQ25cHM1kg6FDgWOIwQl+oxgpBdYGbral1fRNpheOiC5Tgb0zVLiDVCK3taLlhON5DrazySdpJ0vqRFkp6O2/Mk7dSArU4NXLAcpzppfVpvAK4kTG0oD8T3PHCYmf0+FwtbSCt6Wi5YTjfRSE8rrWjdCrwyHj5AWFHnpcCOMe9WM9u3HmOLQLNFywXL6TbyHB7uQehVfdTM+szsIDPrAz4az7+8HkOLQLMd8S5YjpOOtD2tRcBuwCQzW5nIn0BYbOIuM3tFbla2iGb1tFywnG4lz57W5wk9rRPL8k8A1gGfSVmPU4YLluPUR9qe1m+AvYEtCT6tZcAOBJ/WY8BdieJZL3LRMvLuablgOd1Ono74IYYXtqDC/gtFiREg0jTe7uQpWi5YjpPjjHgyXNjCccFynNHgM+JrkEdPywXLcYbJzREvaf8Rzk9PU0+RyGPKgwuW44yetE8Pr5c0ENcsfAFJ20i6BPhR9qa1lqxD07hgOU421OuIvxN4j5n9RdIxwLcJK07TKc73JFkND12wHKcyec7T+iph5Z1XADdJ+j1hPcLJwArgA3VZ2kW4YDlOttSz7uErge8T3kEsTXm4AnifmT2Um4UtZLQ9LRcsx6lN3itM7wpsx7BgGdBPmGTqlOGC5Tj5kPbp4RXAhQTRehI4j/D6zp7AHyR9PTcLM0LSoCRLprzacsFynPyoxxEP8AvCcPDROFycQ3i9p+1nwUsaBFYRFp4FwMxqLsjRyPDQBctx0tPI8BAzGzERlr0/rkL+WOA/gDVp6klctzNwPuGdxdKTyXVVyvYC84DlwHPAbcBJ9bQX6xkEFgAT67jGSikNQ0NDdsoppxhgM2fOtKGhoVTXOU63UvYbS/e7TFUIth/h/D5pG4zlj0waW020CMPRZRXKGjCrzjYHEwK5EvhyimtSi5YLluPUT26iZcM/4t2A9wGfqee6CvXsBwwAhwM31RCtcxMf6rgoYjfG4zVAb6LsYBVxuzye/zxhsdl3AoviuSNHsDOVaLlgOU5j5CpawLcIzvf1JYGJP/71wDFp66lQ74JKokV4SPBUPLcokX904oOemsifCuxeIe1Qoc2PpOmtpREtFyzHaZxGRCtVlAdJHwQ+VOHUmYQniUcBF6epqw6mApPi/r2J/OT+PqUdM1tcrSJJrwD+k7BO41rg1HjqhirlZwAz0hj5xBNPcMkll7jT3XGaRNrQNO8nqOGZwMcT+VfE7T4bXTF6piT2n6myv03Kuh6L2y8AE4H7gJPN7LpKhc3sPOC8NNMiJk+ezM0338zkyZNdsBynCaQVrV3idjYbitbjcbt9ZhYNU00BypcwGxEzexh4a12N1xHlYcqUKSMXchwnE9LOiH8+bieU5e9bdj5Llif2t0jsT6xSJlMs4ygPjuNkQ1rRujVuzy5lSPoIcBGht7MwY7sAFgNPx/1dE/nJ/T/n0C7Q/CXEHMdJR9oZ8dOAX1Y6RRCtN5vZNakbDXG5Sk72S4GDCE8ht4t5K8xstaRzGXaIHwdcG8sfQHCoTzWzB9K2Wy+tWGHacbqJ3F6YNrOrCeKxgiBUpbQC+EA9ghV5LWFot5wgWACbJPL+LebNJqxmDWFW/CMEwQI4PWfB8p6W47QhaR3xmNn3JF1EEJmtCU/k/miJxVuzxswekXQQ8GVgGsGndg/wTTP7Xl7txrYv86eBjtN++MIWNfDhoePkS55LiHUd5cND73U5TnvgPa0a5Blzy3GcDcnUEe84jtMueE+rC5C00Mz2HbmkU47fu8bJ6955T8txnELhotUdnNdqAwqM37vGyeXe+fDQcZxC4T0tx3EKhYuW4ziFwkXLqYikMZL+GNeIvLHV9hQBSS+WdJmkhyQ9L+luSce32q52RtI+km6UtDquTXrCSNe4aDnV+BhhTUsnPZMIATO/BXyKELVkjqSXt9SqNkXSeELUlp0Jf28PAT+Ia6pWxUWrzZG0s6TzJd0laSj2fNZVKdsraZ6k5ZKek3SbpJMaaHMXwnqWnx6t/a2kBfduGbCnmX3ZzL4BXED4jXWEaOVwPw8DdgR+YGbnEMKhA7y3piFpV8Dw1JpEk9eIJPzIriesHt4fr7+x1fehCPeurM4e4G7CAsM7tvpetOP9JPSuDPhoPN4tHl9Vyw7vabU/DwJfAo4A/lSj3OnAS+P+8cBLCGtKAsyW1FsqGH0HViFdDpwA7AV8G+iLl2wqqbRfJJp970plJgCXEYaKJ1mOcd+aTOb3M6KybW1ard6e6vpPt4Cc14gETqPyf8jBVn/+dr938fwUwg96LXBsqz93O99P4G3x+Kvx+JB4fHattj00TWeQ5RqRPwbujIdbA+cAfwNmZmJp+5HlvZsA/J4gYvOBIUnvAu40szurXddh1HM/ryIMI0+QdB8hpDpAzQCfLlqdQWZrRJrZIsLK4Ujqj9lPmNmVo7Cvnclyfc0pBMECmB4ThOFSt4hW6vtpZmskHUn4x3gW8DBhPdJbqYGLVmeQ2RqRScxssEbdnUKW62sO1qivW6jrfprZLQyv+5AKd8R3Bi1dI7Lg+L3Lltzvp4tWZ9DSNSILjt+7bMn9frpotTmSxkmaImkKMC6RPyWmTc1sCLg4ntpD0nRJ2wKnxry1wE+aa3nr8XuXLW1zP1v96NTTiI+WD6byFIRSOiGWy3yCZNGT37vOvJ/e0+oQzOwRwpqUFwCPA6uB2wlPYwZaaVu74/cuW/K+nx4E0HGcQuE9LcdxCoWLluM4hcJFy3GcQuGi5ThOoXDRchynULhoOY5TKFy0HMcpFC5aTiFJRBBdkMhbEPMG66zrYEmnxbRlxnY2ZJNTHQ9N4zjh9ZTZcX8OIfKm06a4aDlNJS4btc7Ci7WZYmYHZ12n03748NCpm+SQR9IbJd2SWJz0qES50qIPcyR9XNJS4HlinCVJB0j6haTHJa2RdK+kz0saV9bemyTdEdtYKOm1I9lVlv8qST+V9GhsZ5mki0rXMNzLAri/vA5Ju0u6UNIj8folks6UNLGsnVcrLDz6vKRFMSqnkzWtfnPcU/ESw4saPEtYIiv5Fv964HWxXCnvibIyWwLTgDVUjgRwWaKtPQgv3CbPrwRWxP0FFewaTOQdWq2dsmsqLuQB/J9EW+VpIbBpLDcZeLLs/DpCwLvCLwzSTsl7Ws5o6AHOJojQOwg/zjGEmOhJXgx8lrDgwR4EsTubEJPpj4T1FTcnrIMHcLikw+L+LGB83J8R6/gGMCGljd+J7QwB74+29gGfhxeGlEl7dzIzmVl/PD4ztvU3wqKrmwH/Gs+9Gjgx7n8s1g1wRrTzI2wYM93JglarpqfiJYZ7J6uBzRP5v4v5zxFigpd6HHeWXb8rlXsuyfRfsex98Xhx4vrNGO59Lahg12CFdn5a4/OclijXn8jvIfSWatl5cSx7XTxeBYxP1LEE72llmryn5YyGx83sucTxg3G7GbBVIr98JZqtU9Rduv4lZXVjZs8ThpwjkWznnhTly3kxsMkIZcrtfMzM1iTOP9RAu04NXLSc0TBZ0maJ49KqwuWi8nzZdY8l9s+yMBzbIBGGghCWlUrWTWwzKYrVSC6gsFuNctWCyj1JGFYCXFLFzjeV2TklPiEtsX0KO506cNFyRsN44AxJW0j6F+B1Mf+PFsdGVbgXuD/unyzpzZI2k7S1pHdJ+hPB7wTwh7h9maT3StqC4I8azwiY2b3A3+Phv5Sul7SDpM8kij6Z2N8rcf0qwpAX4AhJx0t6kaQtJR0u6VfAG8rs3ByYFdv5AFC+BLwzWlo9PvVUvMSw72gF4UneSE8P51So4y2ERQ6q+Yr6Y7nd2fjp4XMEZ/6onx7GMgdWOD8vnntVhc+YTAfHcpWeHg4l8gazuPee3KfljI7HCcOjmwnCci9wjJldP9KFZvZL4PXAL2I9a4AHgCsJQ8OHYrm7gbcBd8UyfyEIXqq188zsV8BrgJ/Ha9YS/GM/TpS5gfB08wGGh4Olc7cC+wEXAo/G6x8GfgOcAtwayz1OEMg/RTvvBd4F3JbGTic9HiPeqZs4IfONwBIbnhrgOE3Be1qO4xQKFy3HcQqFDw8dxykU3tNyHKdQuGg5jlMoXLQcxykULlqO4xQKFy3HcQqFi5bjOIXi/wO2tEauZGn7WAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "################################# Predict metabolome from individual's metagenome.\n",
    "from numpy import array\n",
    "from scipy.sparse import csr_matrix\n",
    "import numpy.matlib\n",
    "from scipy.optimize import minimize, rosen, rosen_der\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "MAX_ID = 2244  # MAX_ID is the maximum of ID labels.\n",
    "\n",
    "def Ain_out(b_real, net):\n",
    "    '''\n",
    "    Ain_out is a function used to create sparse matrices made of metabolites and microbes \n",
    "    where metabolite consumption and production is considered. The matrices created are \"m2b\" and \"b2m\":\n",
    "    (1) m2b is a matrix determines the nutrient splitting among microbes, and\n",
    "    (2) b2m is a matrix determines the byproducts generation.\n",
    "    Both matrices have rows representing bacterial species and columns representing metablolites.\n",
    "    Two matrices are created based on (1) the metabolite consumption and production network which is \n",
    "    encode in \"net\" as a dataframe, and (2) the experimentally measured bacterial abundance \"b_real\".\n",
    "    '''\n",
    "    valid_index = np.where((net.iloc[:,2]==2) | (net.iloc[:,2]==5))[0]\n",
    "    row = net.iloc[valid_index,0]\n",
    "    col = net.iloc[valid_index,1]\n",
    "    data = np.ones((len(valid_index),))\n",
    "    m2b = csr_matrix( (data,(row,col)), shape=(MAX_ID,MAX_ID)).toarray()#.todense()\n",
    "\n",
    "    valid_index = np.where((net.iloc[:,2]==3) | (net.iloc[:,2]==5))[0]\n",
    "    row = net.iloc[valid_index,0]\n",
    "    col = net.iloc[valid_index,1]\n",
    "    data = np.ones((len(valid_index),))\n",
    "    b2m = csr_matrix( (data,(row,col)), shape=(MAX_ID,MAX_ID)).toarray()#.todense()\n",
    "\n",
    "    ########## Normalize the b2m by out_degree\n",
    "    out_degree = b2m.sum(0).copy()\n",
    "    out_degree[out_degree==0]=100\n",
    "    b2m = (b2m / out_degree)\n",
    "\n",
    "    ########## Normalize the m2b by proportion of microbial abundance in each individual\n",
    "    b_real_repmat = numpy.matlib.repmat(b_real[np.newaxis,:], MAX_ID, 1)\n",
    "    m2b = m2b * b_real_repmat\n",
    "    in_degree = m2b.sum(1)\n",
    "    in_degree[in_degree==0]=100\n",
    "    m2b = m2b / numpy.matlib.repmat(in_degree[:,np.newaxis], 1, MAX_ID)\n",
    "    return [m2b, b2m]\n",
    "'''\n",
    "def m2b_multiple_levels(f, m2b, b2m, numLevels_max):\n",
    "    m2b_multiple_levels is a function used to generate matrices involving the calculation of metabolite \n",
    "    byproducts and microbial biomass after several trophic levels/layers. Those matrices are:\n",
    "    (1) m2m_layer is a conversion matrix from the nutrient intake to the metabolite byproducts at a trophic\n",
    "    level or layer.\n",
    "    (2) m2m_total is a conversion matrix from the nutrient intake to a summation of metabolite byproducts at\n",
    "    all trophic levels or layers.\n",
    "    (3) m2b_total is a conversion matrix from the nutrient intake to a summation of all microbial/bacterial \n",
    "    biomass gain at all trophic levels or layers.\n",
    "    Those matrices are computed based on (1) metabolite consumption matrix \"m2b\", (2) metabolite byproduct\n",
    "    generation matrix \"b2m\", (3) byproduct/leakage fraction \"f\", and (4) number of trophic levels/layers in the \n",
    "    simulation \"numLevels_max\".\n",
    "    m2m_layer = np.zeros((MAX_ID, MAX_ID, numLevels_max));  \n",
    "    m2b_total = np.zeros((MAX_ID, MAX_ID));  \n",
    "    \n",
    "    s_step =  np.dot(b2m, m2b.T) # s_step is the conversion matrix of each trophic level/layer\n",
    "    s_step_ii = np.eye(MAX_ID, MAX_ID);\n",
    "    #f_mul = numpy.matlib.repmat(f[np.newaxis,:], MAX_ID, 1)#numpy.matlib.repmat(f, 1, MAX_ID)\n",
    "    f_mul = numpy.matlib.repmat(f, 1, MAX_ID)\n",
    "    \n",
    "    for ii in range(numLevels_max):\n",
    "        # m2b_total is a series made of s_step of each layer\n",
    "        m2b_total = m2b_total + f_mul**(ii) * s_step_ii\n",
    "        m2m_layer[:,:,ii] = f_mul**(ii) * s_step_ii\n",
    "        s_step_ii = np.dot(s_step_ii, s_step);  \n",
    "    m2m_total = m2b_total;\n",
    "    m2b_total = np.dot((1 - f_mul) * m2b.T, m2b_total) # m2b_total has an extra multiplication of m2b and (1-f).\n",
    "    return [m2b_total, m2m_total, m2m_layer]\n",
    "'''\n",
    "i_nonzero_microbes = net['microbes_ID'].unique()\n",
    "i_nonzero_metabolites = net['metabolites_ID'].unique()\n",
    "\n",
    "def array_mult(i_row_a, i_col_a, i_row_b, i_col_b, a, b):\n",
    "    final_array_temp = np.dot(a[np.ix_(i_row_a, i_col_a)], b[np.ix_(i_row_b, i_col_b)])\n",
    "    final_array = np.zeros((MAX_ID, MAX_ID))\n",
    "    final_array[np.ix_(i_row_a, i_col_b)] = final_array_temp\n",
    "    return final_array\n",
    "\n",
    "def m2b_multiple_levels(i_nonzero_microbes, i_nonzero_metabolites, f, m2b, b2m, numLevels_max):\n",
    "    '''\n",
    "    m2b_multiple_levels is a function used to generate matrices involving the calculation of metabolite \n",
    "    byproducts and microbial biomass after several trophic levels/layers. Those matrices are:\n",
    "    (1) m2m_layer is a conversion matrix from the nutrient intake to the metabolite byproducts at a trophic\n",
    "    level or layer.\n",
    "    (2) m2m_total is a conversion matrix from the nutrient intake to a summation of metabolite byproducts at\n",
    "    all trophic levels or layers.\n",
    "    (3) m2b_total is a conversion matrix from the nutrient intake to a summation of all microbial/bacterial \n",
    "    biomass gain at all trophic levels or layers.\n",
    "    Those matrices are computed based on (1) metabolite consumption matrix \"m2b\", (2) metabolite byproduct\n",
    "    generation matrix \"b2m\", (3) byproduct/leakage fraction \"f\", and (4) number of trophic levels/layers in the \n",
    "    simulation \"numLevels_max\".\n",
    "    '''\n",
    "    m2m_layer = np.zeros((MAX_ID, MAX_ID, numLevels_max));  \n",
    "    m2b_total = np.zeros((MAX_ID, MAX_ID));  \n",
    "\n",
    "    s_step = array_mult(i_nonzero_metabolites, i_nonzero_microbes, i_nonzero_microbes, i_nonzero_metabolites, b2m, m2b.T) # s_step is the conversion matrix of each trophic level/layer\n",
    "    s_step_ii = np.eye(MAX_ID, MAX_ID);\n",
    "    #f_mul = numpy.matlib.repmat(f[np.newaxis,:], MAX_ID, 1)#numpy.matlib.repmat(f, 1, MAX_ID)\n",
    "    f_mul = numpy.matlib.repmat(f, 1, MAX_ID)\n",
    "\n",
    "    for ii in range(numLevels_max):\n",
    "        # m2b_total is a series made of s_step of each layer\n",
    "        m2b_total = m2b_total + f_mul**(ii) * s_step_ii\n",
    "        m2m_layer[:,:,ii] = f_mul**(ii) * s_step_ii\n",
    "        s_step_ii = array_mult(i_nonzero_metabolites, i_nonzero_metabolites, i_nonzero_metabolites, i_nonzero_metabolites, s_step_ii, s_step);\n",
    "    m2m_total = m2b_total;\n",
    "    m2b_total = array_mult(i_nonzero_microbes, i_nonzero_metabolites, i_nonzero_metabolites, i_nonzero_metabolites, (1 - f_mul) * m2b.T, m2b_total); # m2b_total has an extra multiplication of m2b and (1-f).\n",
    "    return [m2b_total, m2m_total, m2m_layer]\n",
    "\n",
    "def pred_error(x, b_real, m2b_total, i_intake):\n",
    "    '''\n",
    "    pred_error is a function used to compute the logarithmic error between experimentally measured\n",
    "    metagenome and predicted metagenome computed from the model for a certain nutrient intake. It relies on \n",
    "    (1) x: the nutrient intake, (2) i_intake: IDs of the nutrient intake, (3) m2b_total: a conversion matrix \n",
    "    from thenutrient intake to the total biomass, and (4) b_real: experimentally measured metagenome. The \n",
    "    first three is used to compute the metagenome predicted by the model \"ba_pred\" and compare it with the \n",
    "    experimentally measured metagenome \"b_real\".\n",
    "    '''\n",
    "    x_full = np.zeros((MAX_ID,));\n",
    "    x_full[i_intake] = x;\n",
    "    \n",
    "    ba_pred = np.dot(m2b_total, x_full)\n",
    "    pred_error = (np.log10(ba_pred + 1e-6) - np.log10(b_real +1e-6)) / np.log10(b_real +1e-6);\n",
    "    pred_error = np.sqrt(np.dot(pred_error, pred_error.T)) #np.sqrt(np.sum(pred_error**2))\n",
    "    return pred_error\n",
    "\n",
    "def calc_metabolome(x, m2m_layer, numLevels_max, i_intake):\n",
    "    '''\n",
    "    calc_metabolome is a function used to calculate the metabolome from the fitted nutrient intake from the\n",
    "    model. It relies on (1) x: the nutrient intake, (2) i_intake: IDs of the nutrient intake, (3) m2m_layer: \n",
    "    a conversion matrix from the nutrient intake to the metabolite byproducts at a trophic level or layer, \n",
    "    and (4) numLevels_max: the number of trophic levels/layers in the model. The metabolome in the model is \n",
    "    assumed to be composed of two parts: (1) met_levels: all metabolites in the final trophic level/layer \n",
    "    (which is considered to be reaching the end of the gut because of the finite gut length and gut motility.),\n",
    "    and (2) met_leftover_levels: all unusable metabolites from all previous trophic levels/layers. \n",
    "    '''\n",
    "    i_unused = np.where(np.sum(m2b.T,0) == 0)[0]\n",
    "    met_levels = np.zeros((MAX_ID, numLevels_max));  \n",
    "    met_leftover_levels = np.zeros((MAX_ID, numLevels_max));\n",
    "    \n",
    "    x_full = np.zeros((MAX_ID,));\n",
    "    x_full[i_intake] = x;\n",
    "    \n",
    "    for ii in range(numLevels_max):\n",
    "        met_levels[:,ii] = np.dot(m2m_layer[:,:,ii], x_full)\n",
    "        if ii==0:\n",
    "            met_leftover_levels[i_unused,ii] = x_full[i_unused]\n",
    "        else:\n",
    "            met_leftover_levels[i_unused,ii] = met_levels[i_unused,ii-1]\n",
    "            \n",
    "    return [met_levels, met_leftover_levels]\n",
    "\n",
    "f_byproduct = 0.9\n",
    "f = f_byproduct * np.ones((MAX_ID,1));\n",
    "f[i_selfish] = 0.0;  # The byproduct/leakage fraction f for microbes that don't generate byproducts is set as 0.\n",
    "\n",
    "numLevels_max = 4\n",
    "\n",
    "######## Select an ID of a individual and pull out its experimentally measured metagenome:\n",
    "pa = 3; \n",
    "b_real = np.zeros((MAX_ID,))\n",
    "b_real[thai_metagenome_ID] = thai_metagenome.iloc[:,pa] / np.sum(thai_metagenome.iloc[:,pa])\n",
    "\n",
    "######## Compute matrices involving the metabolite consumption and generation:\n",
    "m2b, b2m = Ain_out(b_real, net)\n",
    "#m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(f, m2b, b2m, numLevels_max)\n",
    "m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(i_nonzero_microbes, i_nonzero_metabolites, f, m2b, b2m, numLevels_max)\n",
    "\n",
    "######## Assign diet\n",
    "x0 = np.ones((len(i_intake),)) / len(i_intake)\n",
    "\n",
    "######## The model is converted into an optimization problem where the nutrient intake is constantly changed to\n",
    "# minimize the logarithmic error between experimentally measured metagenome and predicted metagenome computed \n",
    "# from the model for a certain nutrient intake.\n",
    "fun = lambda x: pred_error(x, b_real, m2b_total, i_intake)\n",
    "bnds = ((0, 100), ) * len(x0)\n",
    "res = minimize(fun, x0, method='SLSQP', bounds=bnds, options={'disp': True, 'maxiter': 1000}, tol=1e-3)\n",
    "#res = minimize(fun, x0, method='Nelder-Mead', bounds=bnds, options={'disp': True})\n",
    "#res = minimize(fun, x0, method='TNC', bounds=bnds, options={'disp': True, 'maxiter': 1000})\n",
    "print(res)\n",
    "\n",
    "######## As long as the optimized nutrient intake is found by using the above optimization solver, the\n",
    "# optimized nutrient intake \"res.x\" is used to generate the predict metagenome and metabolome. They are\n",
    "# visually and statistically compared to the experimentally measured metagenome and metabolome.\n",
    "x_full = np.zeros((MAX_ID,));\n",
    "x_full[i_intake] = res.x;\n",
    "ba_pred = np.dot(m2b_total, x_full)\n",
    "\n",
    "#### Metagenome comparison\n",
    "fig, ax = plt.subplots()\n",
    "ax.loglog(ba_pred+1e-5, b_real+1e-5, 'ko')\n",
    "ax.plot([1e-5, 1], [1e-5, 1],'k-')\n",
    "ax.set_aspect('equal')\n",
    "ax.set_xlabel('predicted')\n",
    "ax.set_ylabel('experimentally observed')\n",
    "ax.set_title('Metagenome comparison')\n",
    "\n",
    "met_levels, met_leftover_levels = calc_metabolome(res.x, m2m_layer, numLevels_max, i_intake)\n",
    "metabolome_measured = np.zeros((MAX_ID,))\n",
    "metabolome_measured[thai_metabolome_ID.values] = thai_metabolome.iloc[:,pa]\n",
    "metabolome_pred = met_levels[:, -1] + met_leftover_levels.sum(1) #np.dot(m2m_total, x_full)\n",
    "#metabolome_pred = np.dot(m2m_total, x_full)\n",
    "i_common = np.where(metabolome_measured * metabolome_pred > 1e-5)[0]\n",
    "metabolome_pred_common = metabolome_pred[i_common] / np.sum(metabolome_pred[i_common])\n",
    "metabolome_measured_common = metabolome_measured[i_common] / np.sum(metabolome_measured[i_common])\n",
    "\n",
    "#### Metabolome comparison\n",
    "fig, ax = plt.subplots()\n",
    "ax.loglog(metabolome_pred_common+1e-5, metabolome_measured_common+1e-5, 'ko')\n",
    "ax.plot([1e-5, 1], [1e-5, 1],'k-')\n",
    "ax.set_aspect('equal')\n",
    "ax.set_xlabel('predicted')\n",
    "ax.set_ylabel('experimentally observed')\n",
    "ax.set_title('Metabolome comparison')\n",
    "\n",
    "print('-------------------------------------------------------------------------------------------------------------')\n",
    "print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:')\n",
    "print(pearsonr(ba_pred[b_real>0], b_real[b_real>0]))\n",
    "print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:')\n",
    "print(pearsonr(metabolome_pred_common, metabolome_measured_common))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find the number of unique microbia and metabolites in the Chia network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "544 221\n"
     ]
    }
   ],
   "source": [
    "uniqueChiaBactIDs = np.sort(net['microbes_ID'].unique())\n",
    "uniqueChiaMetIDs = np.sort(net['metabolites_ID'].unique())\n",
    "print(uniqueChiaBactIDs.shape[0], uniqueChiaMetIDs.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find the number of unique microbia and metabolites in the Thai data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72 81\n",
      "5832\n"
     ]
    }
   ],
   "source": [
    "print(thai_metagenome_ID.shape[0], thai_metabolome_ID.shape[0])\n",
    "print(thai_metagenome_ID.shape[0]*thai_metabolome_ID.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.         0.         0.         0.2030303  0.09090909 0.\n",
      " 0.         0.         0.         0.         5.98020787 0.\n",
      " 0.         0.40854978 0.         0.         0.         0.\n",
      " 0.         0.         0.         0.         0.91525476 0.\n",
      " 0.         0.         0.         0.         0.         0.4789272\n",
      " 0.1        0.         0.         0.         0.         0.1\n",
      " 0.1        0.         0.         0.         0.         6.28751306\n",
      " 0.         0.         0.         0.         0.         0.\n",
      " 0.         0.         0.         0.         0.         0.\n",
      " 0.         0.         0.         0.         0.         0.\n",
      " 0.         0.         0.         4.35636227 0.16364943 0.24604978\n",
      " 0.         0.03448276 6.44104842 0.         0.17938312 0.\n",
      " 0.         0.12539185 0.         0.10114943 0.         0.03448276\n",
      " 0.         0.19698276 0.        ]\n",
      "Consumption connectance is:  0.032578875171467764\n",
      "Production connectance is:  0.030349794238683128\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXkAAAA1CAYAAACk7pueAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAHw0lEQVR4nO2dbYxdRRnHf/9dgRaBkmbVWsO2ikpQEkULCd9KNNEPphCDSFzBNU1MjAnB14BUDZjGhA8mxgQBBb+0hKRGoDEaEl/4QG1JGmhJKFJFqCBCaqUq0Pc+fjhzs+PtvXvP7p7be+/s/5ec7JyZ586Z+Wfuc2fnzHmOIgJjjDFlMjboBhhjjOkfdvLGGFMwdvLGGFMwdvLGGFMwdvLGGFMwdvLGGFMwjTt5SZOSNknaL+mQpN2S1jd9HWOMMb1Rk/vkJa0AdgLv6lC8ISI2NnYxY4wxPWl6Jn8bMw7+euCdwOPp/HuSJhu+njHGmFlozMlLGgM+m06fiYhNEfEK8MOUdwZwTVPXM8YY05smZ/IXAstSem+Wn6cvbfB6xhhjevCWBuuayNL/6ZJ+e7cPS3IQHWOMWQARofa8Jmfyp1TeIf8URy7pS5J2NtgOY4wxiSad/P4sfV6WPreLDQARcU9ErGmwHcYYYxI9l2skTQM/71L8cERcndLPAYeApcA6SfuBR4Btmf2TdRr1cfn+rDHG1OG38YtZy5tck/82lYOHaolmApgCPpPyjgFbGryeMcaYHszFye+LiNWdCiStAr6bTo8CZ2bFrfRtEfHinFtojDFm3jS1Jn8N1T54gJuA+4EDzNxoPQT8oKFrGWOMqclcnPxKSQckHZW0V9Ltks5KZR/J7LZHxFRETABbU95S4N1NNNgYY0x95uLkzwCWp7/vA74DPJzKFrRH3hhjTH+o4+T/DKwHVgNnA1cCr6ayT0hayzz3yIP3yRtjTD/p6eQjYltE3BcR+yLiUEQ8CvwoM7mMee6RT/V7n7wxxvSJOvvkxyLiZFt2Pis/CTwBfC6dvx/YlaUB/g08X7dRvfZ9GmOMqUed5Zqtkm5MLwNZkpZnbsrK/0i1//1YOv+qpBWSpoCLU94DHX4ojDHG9JmeLw2RtAv4UJfi+yNiKtltAL7fwebvwJoUdni26zhAmTHGLIBOAcrqOPl1VEsxa4CVVEs1z1CFOvhJPkNPr/m7EbgIeJ0qrMEtEfG3Og2UtNPr8/WwVvWwTvWxVvUZJa16rslHxFZm9rv3sr0XuHehjTLGGNMMjb/Ie4HcM+gGjBDWqh7WqT7Wqj4jo1WjL/I2xhgzXAzbTN4YY0yDDIWTT9szN0naL+mQpN3pJm7RSJqWFF2Oh9ps1yddDiedNkm6oEOdI6+lpPdK+qmkpyWdTHoc72A3Lumbkp6VdETSy5LulLS8g+0lkrZKek3Sm5K2S7pqIXUOA3PQqts4C0nnt9kWp5WkGyQ9KOn51KdXJf1O0pVtduWNqYgY6AGsAF6i2rXTftw66Pb1ue/TXfodwEOZ3YYuNi8CK0rTEri6Q/uPd7D7WZe+PgUszew+SPVAXifbz8+nzmE55qBVt3EWwPmlawX8aZb+X1fymBoG8e/OxUmOakc6PwpMDrqNfez7dOrnC7PYrEo6RNJlRdKppdldpWlJFSpjI/Ap4PFOjgu4IuvrL6mC5N2c5d2c2T6S8o4AH6OKw/TXlPcv4Oy51jksRx2tkl2rD2t71FekVul7cGv6Pp2bNGu19emSx9SghR8DDqbO7snyr81E+NqgB0gf+z9Nbyf/9UyLa7P8PSnvYNKxSC2BRzs5LuDHWb8uT3njVJFPA3gq5b0NOJHyfp19/lvZ5z89lzqH9eimVSrr6eRL1go4p+1czMzEj5Q8pga9Jn8hsCyl92b5efrS09ecgVE3Vn8njZZRxepfbFqeoktEnKCaTQF8IGn4YWbuPfXSpW6do8wWScfSPZstki7OyorVKiJeb8s6k8rZQvVUPhQ6pgbt5B2HvqKJWP2LTcte/R2n0nQuutStc5SZoHoIcoLqjW47JF2UlbUoXatvAG9N6dYDnEWOqUE7+XnHoS+EJmP1LzYt+6FLyRpuBD4KnEM13lqTiPOAW1J6UWgl6Qbg9nT6B+COVlG3j2TpkRtTg3by845DXwLRbKz+xaZlr/6eAF6rYZfXVbfOkSMiNkTEExHxRkTsA76cFV+W/havlaQvUMXdGgMeA66KiFYE3SLH1KCd/HNUNz9gJvZ8e/rJ09ec04ukTvrnv+qtWP0tOmnUitW/2LQ8RRdJ48B7Ut6eiDhC9W6Dk7ldh3RLl7p1jhQ1xxkUrpWkaeA+Kr/3e+CTEfHfzKTMMTUEd73zbX9TwDv4/21/Fwy6jX3s+6+oonZOAkuAtcArmR5XpLLWFsrtVNsipzKbblsoR1ZLqnsTE+nYltp/PMs7i+a2ux1giLa79Umrr6SxcXkaZ6uAB7N+3V26VsAXmdkR8xtgSQebIsfUMIhfxAM88+z7ri79DmBzZtftYaiXKPNhqLWz6BLAdLJr4sGV69uuPTIP+NTViuolP93K/0H241+qVsALPXRaXeqYGrj4SYRJYDPwT+AwsBtYP+h2nYZ+rwMeAP4CvAm8AeykmnmNtdmuT7ocTjptpsPDTSVoWcdxJbtxqr3Jz1LNqF4G7gSWd6jzEqqQ2QeT1juo1mPb7WrXOQxHHa2ottfeQbWEcIDqv7p9wF3AysWgFfWdfHFjylEojTGmYAZ949UYY0wfsZM3xpiCsZM3xpiCsZM3xpiCsZM3xpiCsZM3xpiCsZM3xpiCsZM3xpiCsZM3xpiC+R981dq8+g0DHgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAOsAAAECCAYAAAAfNiArAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAATS0lEQVR4nO3dfZBcVZ3G8e8zMxAyeUExkkRWHAMbJcqLSBmCKOiypbViaaoslzW8aIUKq1jr+sISS1xfCl8KS3ctLJAsyurGFC6WqdrarXV9QXEtCBQChiVAlCQiQkLAJIRkCJnJb/+4d7oPMz3T3TM9M33Sz6eqq88959y+5/ad35zb574pIjCz9tc13Q0ws8Y4WM0y4WA1y4SD1SwTDlazTDhYzTLR8mCVdLyktZJ2SuqX9BtJK1u9HLNOo1YeZ5W0ALgbOK5G8VUR8YWWLcysw7S6Z/0c1UC9CFgI3FlOf0bS8S1enlnHaFnPKqkL+BNwNPBgRCwp898LfL+s9vGI+Noo8/tUKrNSRGh4Xit71hMoAhVgc5Kfpl/XwuWZdZRWBuu8JP3MKOljW7g8s47S08LPGtFt18gfsasraRWwKs07T+9pYbPM8vDT+MGY5a3sWXcm6blJes4odQCIiDURcUYL22F2WGplsD4C7CnTi5P8NH1vC5dn1lFaFqwRcYjqqO9JklZImg98rMw7CNzSquWZdZpW/mYF+AzwDopjrWuHlX0uIv7Q4uWZdYyWnhQREduBs4B1wNPAAWAjcKnPXjKbmFb3rETEo8CKVn9ubnoWzK+kB7bvGLNu94uOrqQHd+8Zo2by+X3Vk8EGtj3aZOssR77qxiwTDlazTLR8N9gK9XZ9U43u+lpnc89qlgn3rJnyoFLncc9qlgkHq1kmHKxmmXCwmmXCwWqWCQerWSYcrGaZcLCaZcLBapYJB6tZJhysZplwsJplwsFqlom2vOqm67QlHLpv05h1ehb1AfDMqdXbp8zeureSPnRU86vWvXV7JS1V700+0Fcso+u5gWrl3/6+uqyTTxj5YRs2Nr7cJm7rsn/5UgDm3vtEJS+OPKJm3eidUbQv+S6H5ofq91Xvux546+tr5vfsP1i8P1l96ELMnFFJH5g/u5I+aktxy+j/uv0/Knlve9lpYy431b24+I4HNz/S8DxNOfOUSnKgt/g+e2799eQsa5zcs5plwsFqlomWPkx5ItJHPvpZN9aJ0mfdTPYjH81sEjlYzTLhYDXLhIPVLBNteZx1unTPrz6YfXDHkxP6rPTxGfTOrCRj334A1FX9Pxn9/dXl+h7CNgr3rGaZcLCaZcK7wYmJ7vqmmnl8hlkj3LOaZcLBapYJB6tZJhysZplwsJplwsFqlgkHq1kmHKxmmXCwmmXCwWqWCQerWSbqBqukiyWtl7RV0n5JOyT9TNJbhtXrlnSFpIclHZD0uKTrJB0zec036xx1b5gm6SHgVaMU/01E3FzWuxFYWaPO/cDSiOivUZYuxzdMs47Wihum7QauAvqAucAXk7JPA0haRjVQ1wMvBT5ZTp8MfKS5ZpvZcI30rLMj4tlkWhQBPBd4PiJmSLoW+HBZZWlE3CWpG9gFzAHuj4hThn/2sOW4Z7WONuGeNQ3U0pFAd5n+Y/l+elK+uZxvENhS5i2RNAMzG7fxjAZ/AphVpr9Vvs9Lyp+pke4Gag40SVol6e5xtMOsozQVrJIuBj5fTv4cuGaoaLRZknTN/e2IWBMRZzTTDrNO1HCwSroEuKmc51fAuyLiYFm8M6k6N0nPKd8HKX6/mtk4NRSskt4PfLusfyvw9ojYm1S5J0kvLufpBhaVeZsi4sCEW2vWwereME3SB4AbKQL1R8DyiHhuWLV1VEeDV0taBVxKtWdd15rmTq5m7hvcNWtWJX1o374R5RO9b3D6HNXe9XeO3e7FNZ4PyyQ+y9SmRSM962eSem8H+iVF8uqLiDuoDjYtp9gt/lI5fT/w9VY22qwTtfLc4MuAKykO3TwPPAFcD5xb7+wlM6vPz2c1axN+PqvZYcLBapYJB6tZJhysZplwsJplwsFqlgkHq1kmHKxmmXCwmmXCwWqWCQerWSYcrGaZcLCaZcLBapYJB6tZJure1mXanZncG3zDxkqyZ1EfAANbtk3KYtPbsgxs3zGivN5tXay1hm5d08m3qnHPapYJB6tZJnxbF5uwoZ8kMHk/SzqBb+tidpho/wEma3vuTaeGe1azTDhYzTLhYDXLhIPVLBMOVrNMOFjNMuFgNcuEg9UsEw5Ws0w4WM0y4dMNbcJ8Iv/UcM9qlgkHq1kmHKxmmXCwmmXCwWqWCY8G24R5BHhquGc1y4SD1SwTDlazTDQdrJLOkxTJ6+xh5Ssl/UbSc5J2Slor6eWta7JZZ2oqWCUdAVw7RvlVwI3AKcAMYB6wArhd0oIJtNOs4zXbs34UeDWwf3iBpFcA/1hO3gksBC4qp/8M+Oz4mmhm0ESwSnoZ8GngSeBfalR5D3BEmf5aRGyPiLXAg2XeBZL8G9lsnJoJnq8Cs4Ergd01yk9P0ptrpI8GXtlU68ysoqFglXQOcAFwO/CdUarNS9LPjJI+tsZnr5J0dyPtMOtkdc9gktQDfAMYBC6PiJBGPDMHoGbmsPwRT8GKiDXAmvTBVNOle371f0m6juN5Pmv6fFd6Z1aSsa/4ua+u6v/J6O+vpvuOKz7zvk2VvP95/L5K+k2XX1Z85Po7x1iTF9q/fGnN/Nlb9wLw7CvnVJv6RLUtQ8/DHXjr62vO37P/YPH+ZPX/ccycUUkfmD+7kj5qy87is7Y9OmZbn1q1rJI+9q49lXT6fUyK5DnAA73Fr7mh9Ruh3vdy669rfu6YkmcPj6aR0w3fDbwW+E8ASacB6cjuiZK2AzuTvLlJek6STuuYWRMa2Q0e+vd4PnBv+bosKb+J4nDNPUne4hrpPcDW8TXTzOo+n1XS+ykCciy3ARcDv6MYEd4ALAf+Alhb1rkhIv52jOX4+azW0Sb8fNaI+NeIUPoCPpdUeVNEnBsRjwKfL/POBJ6gGqh/xMdZzSakpcc9I+Jq4FJgI3AAeBpYB5wVEdtbuSyzTlN3N3iqeDfYOt2Ed4PNrD04WM0y4WA1y4SD1SwTDlazTDhYzTLhYDXLhIPVLBMOVrNMOFjNMuFgNcuEg9UsEw5Ws0w4WM0y4WA1y4SD1SwTDlazTDhYzTLhYDXLhIPVLBMOVrNMOFjNMuFgNcuEg9UsEw5Ws0w4WM0y4WA1y4SD1SwTDlazTDhYzTLhYDXLhIPVLBMOVrNMOFjNMuFgNcuEg9UsEw5Ws0w4WM0y4WA1y4SD1SwTTQWrpIsk3S5pr6R9kh6SdE1S3i3pCkkPSzog6XFJ10k6pvVNN+ssDQerpG8A3wWWAbOBXuBVwPuSajcA1wCLgSOBhcAHgV9ImtmiNpt1pIaCVdL5wOXl5C0UwdgLvBb4YllnGbCyrLMeeCnwyXL6ZOAjrWmyWWdqtGf9u/J9G3BhRPw2Ivoj4oGIuK4sS3vYL0fEU8BXgL01ys2sSXWDVVI38MZy8jFgvaTdkv4k6XuSFpZlpyezbQaIiEFgS5m3RNKMFrXbrOP0NFDnJRS7vABnDyt7H3CGpNcB85L8Z2qku4FjgCfG0U6zjtfIbvARw6YvAeZSDCZB8ft1BaBR5k/zY0ShtErS3Q20w6yjNRKsu6gG2a6I+G5E7AWuS+qcCuxMpucm6Tnl+2D5WS8QEWsi4ozGm2zWmeoGa0TsBx4emhylWj9wTzK9GCq/dxeVeZsi4sA422nW8RodDb65fD9G0sWSZgMfSspvA9Yl06slzQOuoNqzpuVm1iRFjNZZJpWK4LwLOKlG8U+At0VESLqR6rHW1P3A0ojoH2MZlYacp/fUbZPZ4ean8YNKOiJGjAE11LNGxLPAOcAaYAdwkOKQzNXAO6Ma8ZcBV1IcunmeYuT3euDcsQLVzOprqGedCu5ZrdPV61kbOc465bpOW8Kh+zYVE2eeUi3YsLGS7FnUB8DAlm0vmG/IoaOaX7Xurdsraan6XQ30zS8+/7mBauXf/r66rJNPGPlhSVvrLvdFR1fSg7v3NDzfeOxfvrSS7l1/Z0s+c2hbwAu3xwvq9B1flG97dFzL6F5cfMeDmx9peJ6nVi0bs3zemjuqE8nf2UBvcbSy59Zfjzn/wFtfX0nXq9sKvkTOLBPeDU50zz82bU8lPbB9x4i6XbNmVdKH9u2b3IZZR2jJAJOZTT8Hq1km2nKAaboM7niy4br1dn17FsyvpNPd6KFdbXVV/09Gf/Wo1mQPMFm+3LOaZcLBapYJ7wZPklojyNDcrrZZyj2rWSbcs46Tj7PaVHPPapYJB6tZJtp/Nzg5wfp3f91bSb/62uKk+3Y4kf8FFxuUerYlA0y91fubx779xeePcpw1+o4DqF7IkJnu17yqkh584OExataWnnyfnmhf70T+3/3TmSPyXvTgaLcFK+w+qXqq7Ykf3TCyQo3tCjR1kUYruWc1y4SD1SwTvurGrE34qhuzw0T7DzAdZnwiv42Xe1azTDhYzTLh3eAp5hP5bbzcs5plwsFqlgkHq1kmHKxmmfAA0zj5etbmTPSO/Oae1SwbDlazTHg3eJy869sc7/5OnHtWs0w4WM0y4d3gUYz2+IshHg22qeae1SwTDlazTHg3eBSjPf5iiHd9baq5ZzXLhHvWcfIAk00196xmmXCwmmXCu8Hj5F1fm2oN96ySLpZ0h6SnJO2T9LCkL0l6cVKnW9IVZdkBSY9Luk7SMZPTfLPO0VDPKukK4Jph2YuB1cBbgKGnAt0ArEzqLAQ+CJwtaWlE9GNm49Joz3ph+T4InAPMA+4q85ZKWiJpGdVAXQ+8FPhkOX0y8JGJN7eqZ1EfPYv6WvmRdM8/tvKqp2vWrMrLbCo0GqyD5fv2iPhlRDwN/CQpnwm8L5n+ckQ8BXwF2FvmpeVm1qRGB5jWANcDCyW9GXgA+Muy7HHg/4DTk/qbASJiUNIW4FRgiaQZEXGgFQ1Pn8vaKs3c09cDTDbVGupZI+KbwN8DAm4DngLeANwLvKMMwHnJLM/USHcDIwaaJK2SdHfzTTfrMBFR90WxC3sQiGGvx4FLyjqbk/yuZN7/TfIXjLGM4Z/tl18d+6oVI3V7VkldwLUUu8x/BJYAc4HvUIz23iTpDGBnMtvcJD2nfB8EdtVbnpnV1shu8LFUd19/GREPRsReYF2ZJ4rDN/ck8yyG4rgrsKjM29Sq36tmnaiRAaZdwHPAUcCbJb2aoodNR3d3A78CPlxOr5a0CriUas+6jjGkT3qWdHdEnNHQGmTicFwn8HpNpbo9a9kbXl9OHgc8SDFodEmZtwP4YUTcAXyrzFtOsVv8pXL6fuDrLWqzWUdq9DjrFRSjwfcC+4EBit7134CzyuOuAJcBV1IMNj0PPEER6Oc2efbSmibq5uJwXCfwek0ZlSOxZtbmfImcWSbaJlglHS9praSdkvol/UbSyvpzTr/yiqT1krZK2i9ph6SfSXrLsHrZXpUk6TxJkbzOHla+stxmz5XbcK2kl09Xexsh6SJJt0vaW15J9pCka5Ly9tpejZwUMdkvYAHwGLUPEH9qutvXQPsfGqXtAVyQ1LtxlDobgZnTvR5jrN8RFAOLaZvPTsqvGmW9/sAYJ8JM8zp9Y5Q2P9au22vav7TyS7kh+SIuLIN3Qzn9PHD8dLexTvs3AJ8CXkFxqOoLyfo8UNZZluT9kOL0zNVJ3urpXo8x1u8fyjbuGx6s5To/X+ZtKLfdhUm9b053+2usz/lJ+/4d+HOKi1FeA3yoXbdXO3xxXRTHaYPixImh/PcmX8zHprudddZh9rBpAXvKth8o865N1ucNZV43xWGwADZO93qMsm4vo7hyagfwzzWC9eNJ3nuT+TaVebtJTj9thxfw47JtW4EjR6nTdturHX6zngAcXaY3J/lp+nVT15zmRcSzw7KOpNiwUBziglGuSgK2lHlLJM2YtEaO31eB2RSH5HbXKB+xXsPSRwOvnJymNa88q+6N5eRjwHpJuyX9SdL3JC0sy9pue7VDsNa7WgeKUx5z8glg6Kr0oRNFxnVV0nSSdA5wAXA7xbngteS2/V4C9Jbps4G/oviH8mKKs/J+IamXNtxe7RCsaiA/m4PBki4GPl9O/pzq7XCyWk9JPRSDMIPA5VHuB9aq2kB+26wXxWBZ6hKKC09uKKcXAytow/Vqh2Ctd7XO8DptS9IlwE0U3+uvgHdFxMGyOLerkt4NvBb4bwBJp1EMHg05UdKJ5Lf9dlENsl0R8d0oLky5LqlzKm24vdohWB+hGIyB8mqdGul7p6454yPp/cC3Kb7TW4G3l38EQ3K7Kml2+X4+xfd/L8XppENuoji0MWK9hqX3UAzktIWI2A88PDQ5SrV+2nB7TXuwRsQh4Pvl5EmSVkiaD3yszDsI3DItjWuQpA9Q/DbtAn5EcfeM4fd9Sa86Wi1pHsU51w1dldTGbqHYRgAflbRA0grgpDLv5nIbt5Oby/djyhNaZgMfSspvox2313QPo5c/hXI/KWLbKG0fevWV9drqIPs41vOzSZsbOSniMdrwpAiKvYZNo7T5x1TPmW+r7TXtPStARGwHzqL4b/U0cIDiC7k0Ir4wnW1rsVZdldRWIuJqimuXN1Jsu6cptuVZ5bZtK1EcajuH4sqaHRR7BluAq4F3RhmptNn28lU3Zploi57VzOpzsJplwsFqlgkHq1kmHKxmmXCwmmXCwWqWCQerWSYcrGaZ+H8cGDgSzBSmtgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "b_real = np.zeros((MAX_ID,))\n",
    "b_real[thai_metagenome_ID] = 1\n",
    "[m2b, b2m] = Ain_out(b_real, net)\n",
    "\n",
    "plt.figure()\n",
    "plt.imshow(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)])\n",
    "\n",
    "#print(np.sum(m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)], 0))\n",
    "#print(np.sum(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)], 0))\n",
    "#print(np.sum(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)], 1))\n",
    "\n",
    "print('Consumption connectance is: ',np.sum(m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0) / (thai_metagenome_ID.shape[0]*thai_metabolome_ID.shape[0]))\n",
    "print('Production connectance is: ',np.sum(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0) / (thai_metagenome_ID.shape[0]*thai_metabolome_ID.shape[0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(81,)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)], 1).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.43152099376378306, 0.00859394527538631)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6145152231366302, 0.006656010659883466)\n",
      "1\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9112067409061099, 2.676585526059427e-12)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.4113792818925748, 0.1276505214190439)\n",
      "2\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8057777798261692, 6.090867662537397e-10)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8278419141602508, 2.2262172323199077e-05)\n",
      "3\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8070143454797384, 1.233742065563012e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9261608455145216, 7.237895745818847e-07)\n",
      "4\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8566415391611938, 9.068018754287532e-13)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7369346867551363, 0.00031914384838243763)\n",
      "5\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8764596305027472, 1.176759947848287e-12)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7438350659516773, 0.000261282106717019)\n",
      "6\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8434906521669271, 4.844445629631639e-09)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8682763212574451, 2.9901853330918224e-06)\n",
      "7\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.768614715960534, 4.3930169827311533e-08)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.5886518727901788, 0.008014139338306539)\n",
      "8\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.6029150017545283, 3.8341414389764674e-05)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9240058358869017, 3.20489810620613e-07)\n",
      "9\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9848762132409402, 8.973626519128684e-30)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.3937822452337326, 0.10591046127225148)\n",
      "10\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.6770318957163854, 0.00020167135473506354)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7069156022828267, 0.0032096698836194338)\n",
      "11\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8124288019859581, 1.5183462691232887e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9259167532089438, 2.6954958808518464e-07)\n",
      "12\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9943678473451798, 7.646952546659849e-29)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.780977386554358, 7.921242300610871e-05)\n",
      "13\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5841956419945263, 4.8691702438320474e-05)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.13971567757066947, 0.5803097262022635)\n",
      "14\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.6239148229869599, 0.001121983611733732)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.5723405694199545, 0.040945358510254264)\n",
      "15\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9807066535028092, 5.677015733455686e-25)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.5218763447628779, 0.021913477883350994)\n",
      "16\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5819918418418436, 0.00010175736902473167)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.37974949412310144, 0.14683771327709297)\n",
      "17\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9746106145804913, 2.1348542362095522e-20)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.5706748327879112, 0.010720897769295499)\n",
      "18\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.6997419416540537, 5.049376448543155e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.4762809129853961, 0.04570150423727229)\n",
      "19\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9035114041071554, 1.9445417502010516e-14)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8887064701241241, 3.698809479352782e-07)\n",
      "20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7507381573067047, 3.705869637806499e-08)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7481810481404969, 0.00022961977934424403)\n",
      "21\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7417394368724527, 3.425172140833967e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.3091425229951885, 0.21193000537974097)\n",
      "22\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9037543769551467, 1.8642991479601248e-14)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6634932667345975, 0.0026821711655941865)\n",
      "23\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8745307715266151, 1.3906764587789068e-11)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.5334321068766995, 0.022620459491327804)\n",
      "24\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9707255672320361, 6.808371186371102e-24)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8711739333865993, 1.2056524789302664e-06)\n",
      "25\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.41428676280918, 0.008740699034324347)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8882585201874367, 3.8212551360148987e-07)\n",
      "26\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9883405037003169, 2.6333000466038895e-20)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6770517155015756, 0.0014521033782874232)\n",
      "27\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8237660925465249, 3.8933635067376696e-09)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9508238293802332, 4.4320516491491706e-10)\n",
      "28\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9001875642750636, 1.470494025182507e-14)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.668122696600129, 0.004671686065339234)\n",
      "29\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.742239673065754, 6.363367707189242e-08)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7802245115381305, 0.00013368334995066182)\n",
      "30\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.1793759673381802, 0.325937027520048)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8694060867184301, 5.843948210220423e-06)\n",
      "31\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.6426817149709767, 3.372976923444928e-06)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.506524155414137, 0.03194555183011209)\n",
      "32\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8851960057192626, 1.6659455347185125e-12)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6376274492738078, 0.004417004334158609)\n",
      "33\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9963424717045144, 5.472772387056375e-39)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.4887389757431891, 0.06450358516068357)\n",
      "34\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5424522493208884, 0.006170479889706101)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.09045314914457842, 0.8169821133476196)\n",
      "35\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.35580038311176876, 0.0456544519606532)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8885008207132046, 9.606127801700565e-06)\n",
      "36\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.90320103089404, 2.710400295562344e-13)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6751701057418293, 0.0021085344108482105)\n",
      "37\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.13807265630118215, 0.41510029976854496)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.31190268236760343, 0.20766514123178068)\n",
      "38\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-248-b16c8a4067ba>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0mfun\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpred_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb_real\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm2b_total\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi_intake\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m     \u001b[0mbnds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mminimize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'SLSQP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbounds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbnds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moptions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'maxiter'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m1000\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtol\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1e-3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m     \u001b[0;31m######## As long as the optimized nutrient intake is found by using the above optimization solver, the\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/_minimize.py\u001b[0m in \u001b[0;36mminimize\u001b[0;34m(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)\u001b[0m\n\u001b[1;32m    606\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mmeth\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'slsqp'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m         return _minimize_slsqp(fun, x0, args, jac, bounds,\n\u001b[0;32m--> 608\u001b[0;31m                                constraints, callback=callback, **options)\n\u001b[0m\u001b[1;32m    609\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mmeth\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'trust-constr'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    610\u001b[0m         return _minimize_trustregion_constr(fun, x0, args, jac, hess, hessp,\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/slsqp.py\u001b[0m in \u001b[0;36m_minimize_slsqp\u001b[0;34m(func, x0, args, jac, bounds, constraints, maxiter, ftol, iprint, disp, eps, callback, **unknown_options)\u001b[0m\n\u001b[1;32m    421\u001b[0m             \u001b[0;31m# Compute the derivatives of the objective function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    422\u001b[0m             \u001b[0;31m# For some reason SLSQP wants g dimensioned to n+1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 423\u001b[0;31m             \u001b[0mg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfprime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0.0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    424\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    425\u001b[0m             \u001b[0;31m# Compute the normals of the constraints\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py\u001b[0m in \u001b[0;36mfunction_wrapper\u001b[0;34m(*wrapper_args)\u001b[0m\n\u001b[1;32m    324\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mwrapper_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m         \u001b[0mncalls\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwrapper_args\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mncalls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/slsqp.py\u001b[0m in \u001b[0;36mapprox_jacobian\u001b[0;34m(x, func, epsilon, *args)\u001b[0m\n\u001b[1;32m     61\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m         \u001b[0mdx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mepsilon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 63\u001b[0;31m         \u001b[0mjac\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mdx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mf0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mepsilon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     64\u001b[0m         \u001b[0mdx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0.0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py\u001b[0m in \u001b[0;36mfunction_wrapper\u001b[0;34m(*wrapper_args)\u001b[0m\n\u001b[1;32m    324\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mwrapper_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m         \u001b[0mncalls\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwrapper_args\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mncalls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-248-b16c8a4067ba>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m     18\u001b[0m     \u001b[0;31m# minimize the logarithmic error between experimentally measured metagenome and predicted metagenome computed\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m     \u001b[0;31m# from the model for a certain nutrient intake.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m     \u001b[0mfun\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpred_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb_real\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm2b_total\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi_intake\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m     \u001b[0mbnds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m     \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mminimize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'SLSQP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbounds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbnds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moptions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'maxiter'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m1000\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtol\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1e-3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-247-7d30942ee1e7>\u001b[0m in \u001b[0;36mpred_error\u001b[0;34m(x, b_real, m2b_total, i_intake)\u001b[0m\n\u001b[1;32m    124\u001b[0m     \u001b[0mx_full\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi_intake\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    125\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 126\u001b[0;31m     \u001b[0mba_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mm2b_total\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx_full\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    127\u001b[0m     \u001b[0mpred_error\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog10\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mba_pred\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1e-6\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog10\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb_real\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;36m1e-6\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog10\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb_real\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0;36m1e-6\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    128\u001b[0m     \u001b[0mpred_error\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msqrt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpred_error\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_error\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#np.sqrt(np.sum(pred_error**2))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "corr_list = np.zeros((41,2))\n",
    "diet_list = np.zeros((41, 20))\n",
    "for pa in range(41):\n",
    "    print(pa)\n",
    "    ######## Select an ID of a individual and pull out its experimentally measured metagenome:\n",
    "    b_real = np.zeros((MAX_ID,))\n",
    "    b_real[thai_metagenome_ID] = thai_metagenome.iloc[:,pa] / np.sum(thai_metagenome.iloc[:,pa])\n",
    "\n",
    "    ######## Compute matrices involving the metabolite consumption and generation:\n",
    "    m2b, b2m = Ain_out(b_real, net)\n",
    "    #m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(f, m2b, b2m, numLevels_max)\n",
    "    m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(i_nonzero_microbes, i_nonzero_metabolites, f, m2b, b2m, numLevels_max)\n",
    "\n",
    "    ######## Assign diet\n",
    "    x0 = np.ones((len(i_intake),)) / len(i_intake)\n",
    "\n",
    "    ######## The model is converted into an optimization problem where the nutrient intake is constantly changed to\n",
    "    # minimize the logarithmic error between experimentally measured metagenome and predicted metagenome computed \n",
    "    # from the model for a certain nutrient intake.\n",
    "    fun = lambda x: pred_error(x, b_real, m2b_total, i_intake)\n",
    "    bnds = ((0, 100), ) * len(x0)\n",
    "    res = minimize(fun, x0, method='SLSQP', bounds=bnds, options={'maxiter': 1000}, tol=1e-3)\n",
    "\n",
    "    ######## As long as the optimized nutrient intake is found by using the above optimization solver, the\n",
    "    # optimized nutrient intake \"res.x\" is used to generate the predict metagenome and metabolome. They are\n",
    "    # visually and statistically compared to the experimentally measured metagenome and metabolome.\n",
    "    x_full = np.zeros((MAX_ID,));\n",
    "    x_full[i_intake] = res.x;\n",
    "    ba_pred = np.dot(m2b_total, x_full)\n",
    "\n",
    "    met_levels, met_leftover_levels = calc_metabolome(res.x, m2m_layer, numLevels_max, i_intake)\n",
    "    metabolome_measured = np.zeros((MAX_ID,))\n",
    "    metabolome_measured[thai_metabolome_ID.values] = thai_metabolome.iloc[:,pa]\n",
    "    metabolome_pred = met_levels[:, -1] + met_leftover_levels.sum(1) #np.dot(m2m_total, x_full)\n",
    "    i_common = np.where(metabolome_measured * metabolome_pred > 1e-5)[0]\n",
    "    metabolome_pred_common = metabolome_pred[i_common] / np.sum(metabolome_pred[i_common])\n",
    "    metabolome_measured_common = metabolome_measured[i_common] / np.sum(metabolome_measured[i_common])\n",
    "\n",
    "    corr_list[pa, 0] = pearsonr(ba_pred[b_real>0], b_real[b_real>0])[0]\n",
    "    corr_list[pa, 1] = pearsonr(metabolome_pred_common, metabolome_measured_common)[0]\n",
    "    diet_list[pa, :] = res.x\n",
    "\n",
    "    print('-------------------------------------------------------------------------------------------------------------')\n",
    "    print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:')\n",
    "    print(pearsonr(ba_pred[b_real>0], b_real[b_real>0]))\n",
    "    print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:')\n",
    "    print(pearsonr(metabolome_pred_common, metabolome_measured_common))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'corr_list' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-b8531fcf5bfb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmean\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcorr_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'corr_list' is not defined"
     ]
    }
   ],
   "source": [
    "np.mean(corr_list[:,1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute an average diet and stick to the average diet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "average_diet = np.mean(diet_list, 0).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.42036613797943234, 0.010689786637539523)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.62386532243047, 0.005659898402854842)\n",
      "1\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7126071038089303, 9.963251587243866e-06)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.484135857802976, 0.06744597430576534)\n",
      "2\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7733913751871969, 7.848840060521874e-09)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8083191761420259, 4.923706812540085e-05)\n",
      "3\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.4760368655237572, 0.009045865330860424)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9177222058690142, 1.4324267783070164e-06)\n",
      "4\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.48232415612623447, 0.0014056435732095164)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7474139489605269, 0.0002349585641129516)\n",
      "5\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.35181121638858853, 0.03273433757931506)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7617416924755178, 0.00015087745985901057)\n",
      "6\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.33476817587339974, 0.07056472900202475)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7096425250736635, 0.0009713206179698086)\n",
      "7\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7153779137359143, 9.425157184051482e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.5986879549358539, 0.006763355400342218)\n",
      "8\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.3380233112310282, 0.03290047713995)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8514182022053818, 2.867221475285767e-05)\n",
      "9\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9147923107052431, 3.8418916391319426e-16)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.47730744701757033, 0.04517118107656335)\n",
      "10\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.1935099959210756, 0.35402143595035884)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6462096935179318, 0.009246097894552632)\n",
      "11\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.23411515100304314, 0.2304919077152078)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8623149606645011, 1.734143602867544e-05)\n",
      "12\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8935084688632714, 3.047547425067999e-11)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7692281508133291, 0.00011825943195743666)\n",
      "13\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5647330236451418, 9.78093144505119e-05)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.14768360015971999, 0.5586766532125774)\n",
      "14\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5117338385620533, 0.010584368231491571)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6017377604176036, 0.02957184208609182)\n",
      "15\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8587442745413264, 4.1537132732768064e-11)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6259553531884094, 0.0041449390670032045)\n",
      "16\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.4629405826368439, 0.0030008111989658207)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.34098649689782357, 0.19619788385458353)\n",
      "17\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7914148085575279, 1.1522204362358949e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.695034850034099, 0.0009561585012852432)\n",
      "18\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.585171879346478, 7.31090967229608e-05)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.4448974802603517, 0.06431203368594292)\n",
      "19\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5241352535365085, 0.0008695089071135613)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8817463791299204, 6.044284617445063e-07)\n",
      "20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.734437993958225, 1.0265883909519364e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7883446066430754, 6.0849875972201344e-05)\n",
      "21\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5967066097312398, 0.000154665141161594)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.2784663075193508, 0.26317017701445533)\n",
      "22\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7799636575925069, 1.2685494993613567e-08)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7665099439476903, 0.00020700720953538434)\n",
      "23\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7293345056271453, 9.940509182795943e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6580721169073674, 0.0029890418411347083)\n",
      "24\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.6189096812097703, 3.4419653712988415e-05)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8308913605633104, 1.0545551138467908e-05)\n",
      "25\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.26846799280839784, 0.09842682323181544)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9045353719225789, 1.061259030727785e-07)\n",
      "26\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.8404580533485292, 1.4607811079258093e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7051332308397419, 0.0007464182900948721)\n",
      "27\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.17665355330515872, 0.3253952927359432)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.9054762739356848, 9.787783202760103e-08)\n",
      "28\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.2364605686854781, 0.15291537125327476)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6671076926636287, 0.004758013329735081)\n",
      "29\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7000980279573705, 7.025888463992294e-07)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8222722561549028, 2.818953633237942e-05)\n",
      "30\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(-0.01322279001344731, 0.9427400345432524)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.674053237436849, 0.0030055446296023813)\n",
      "31\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.38073025941517624, 0.011778714016097392)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.49267467188337954, 0.037778623730549216)\n",
      "32\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.6137019846424054, 8.832944396298636e-05)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7434186978749331, 0.0004064461453619805)\n",
      "33\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.9148565351270835, 2.389047716160014e-15)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.5725229420047429, 0.025713842473122604)\n",
      "34\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.3756010858552867, 0.07049492038763036)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.06616025348734528, 0.8657101026934052)\n",
      "35\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.12032018976221615, 0.511860380832155)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.8910670790662938, 8.309871310876393e-06)\n",
      "36\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.7814814036412806, 4.892794820874131e-08)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7592602133517512, 0.00025786070133716613)\n",
      "37\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.051358137603275233, 0.7627463758547426)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.36430939095376774, 0.13719780667756284)\n",
      "38\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.3526108634871809, 0.01751627325748253)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7177976024478236, 0.0017417979297138723)\n",
      "39\n",
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.5019807165768724, 0.0013221765438851806)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.7365836427875883, 0.0004898067528818525)\n",
      "40\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------------------------------------\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:\n",
      "(0.4440625130557008, 0.010895447903684055)\n",
      "(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:\n",
      "(0.6548250043004457, 0.0023455075508009293)\n"
     ]
    }
   ],
   "source": [
    "#%%heat\n",
    "corr_list_aveDiet = np.zeros((41,2))\n",
    "for pa in range(41):\n",
    "    print(pa)\n",
    "    ######## Select an ID of a individual and pull out its experimentally measured metagenome:\n",
    "    b_real = np.zeros((MAX_ID,))\n",
    "    b_real[thai_metagenome_ID] = thai_metagenome.iloc[:,pa] / np.sum(thai_metagenome.iloc[:,pa])\n",
    "\n",
    "    ######## Compute matrices involving the metabolite consumption and generation:\n",
    "    m2b, b2m = Ain_out(b_real, net)\n",
    "    m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(f, m2b, b2m, numLevels_max)\n",
    "\n",
    "    ######## As long as the optimized nutrient intake is found by using the above optimization solver, the\n",
    "    # optimized nutrient intake \"res.x\" is used to generate the predict metagenome and metabolome. They are\n",
    "    # visually and statistically compared to the experimentally measured metagenome and metabolome.\n",
    "    x_full = np.zeros((MAX_ID,));\n",
    "    x_full[i_intake] = average_diet.copy();\n",
    "    ba_pred = np.dot(m2b_total, x_full)\n",
    "\n",
    "    met_levels, met_leftover_levels = calc_metabolome(average_diet.copy(), m2m_layer, numLevels_max, i_intake)\n",
    "    metabolome_measured = np.zeros((MAX_ID,))\n",
    "    metabolome_measured[thai_metabolome_ID.values] = thai_metabolome.iloc[:,pa]\n",
    "    metabolome_pred = met_levels[:, -1] + met_leftover_levels.sum(1) #np.dot(m2m_total, x_full)\n",
    "    i_common = np.where(metabolome_measured * metabolome_pred > 1e-5)[0]\n",
    "    metabolome_pred_common = metabolome_pred[i_common] / np.sum(metabolome_pred[i_common])\n",
    "    metabolome_measured_common = metabolome_measured[i_common] / np.sum(metabolome_measured[i_common])\n",
    "\n",
    "    corr_list_aveDiet[pa, 0] = pearsonr(ba_pred[b_real>0], b_real[b_real>0])[0]\n",
    "    corr_list_aveDiet[pa, 1] = pearsonr(metabolome_pred_common, metabolome_measured_common)[0]\n",
    "\n",
    "    print('-------------------------------------------------------------------------------------------------------------')\n",
    "    print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:')\n",
    "    print(pearsonr(ba_pred[b_real>0], b_real[b_real>0]))\n",
    "    print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:')\n",
    "    print(pearsonr(metabolome_pred_common, metabolome_measured_common))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.hist(corr_list_aveDiet[:,1]);\n",
    "print(np.mean(corr_list_aveDiet[:,1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5205970094280451\n"
     ]
    }
   ],
   "source": [
    "print(np.mean(corr_list_aveDiet[:,0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Consumption connectance is:  0.032578875171467764\n",
      "Production connectance is:  0.030349794238683128\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAOsAAAECCAYAAAAfNiArAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAATYUlEQVR4nO3de5BcZZ3G8e8zExIScjEYbgIBlI0hIjeDgZgV2GXLrGIpVZbFEi5aUGG9lK4Xlljieim8YWmthQWSRVldTOFimX+2VtY7LsU1RTAQIFkFJOESAyTkNuQy89s/zpnu15me6e6Zzky/6edT1dXvec97ut/TJ0/eM+fSrYjAzNpf13h3wMwa47CaZcJhNcuEw2qWCYfVLBMOq1kmWh5WSbMl3SZps6QeSb+XdEWr38es06iV51klHQmsAo6uMfvaiPhyy97MrMO0emT9ItWgXgocBdxfTn9e0uwWv59Zx2jZyCqpC3gZmAE8HhHzyvr3Az8um30qIr41xPK+lMqsFBEaWNfKkfUNFEEFWJ/Up+XTW/h+Zh2llWGdlZS3DVE+vIXvZ9ZRJrTwtQYN2zXqB+3qSloKLE3rztf7Wtgtszz8Mn4y7PxWjqybk/L0pDxtiDYARMTyiJjfwn6YHZBaGdY/Aq+U5TlJfVpe3cL3M+soLQtrRPRRPep7kqQlko4APlnW7QXuaNX7mXWaVv7NCvB54F0U51pvGzDvixGxocXvZ9YxWnpRRES8ACwEVgAvAbuBNcCVvnrJbHRaPbISEc8AS1r9urmZcMJxlfK+p/40bNvuI6pntHo3/bmh1+86bV6l3PfwY032znLku27MMuGwmmWi5bvBVqi365tqdNfXOptHVrNMeGTNlA8qdR6PrGaZcFjNMuGwmmXCYTXLhMNqlgmH1SwTDqtZJhxWs0w4rGaZcFjNMuGwmmXCYTXLhMNqlom2vOtmz+IzmXjng8O26TplLgAb/v7QSt0RD+6ulPsmNv//0JS1z1cnuqvL75p7RPGee/oqdZNWP1kp9yw4cdBr1et/qpmvdXlx6dkAHP7AK5W6fVMn1mz76mGTAJiy8v5K3baLz6qUD365t6G+psukJvQU39k+bf3WSt3emZMr5R2zD66UX/No0eZnd95eqXvH604b9n1Tmn8yALHq0YaXacaexWdWyvsmF9s+/dzagUdWs0w4rGaZaOmPKY9G+pOP/q0b60Tpb93s7598NLP9yGE1y4TDapYJh9UsE215nnW8TDj2mEp534aNo3ut5Ocz+mYcUil3bd1RFqr/T8bOXZWyv0PYhuKR1SwTDqtZJrwbnBjtru9fvNYQP5/RV7PWrD6PrGaZcFjNMuGwmmXCYTXLhMNqlgmH1SwTDqtZJhxWs0w4rGaZcFjNMuGwmmWiblglXSZppaSnJO2StEnSrySdN6Bdt6SrJa2TtFvSc5JulHToUK9tZo2r+4Vpkp4A3jjE7H+IiNvLdrcAV9Ro8wiwICJ66ryPvzDNOlorvjBtK3AtcDwwHfhKMu9zAJLOphrUlcBhwGfK6TcDH2+u22Y2UCMj69SI2JFMiyLA04E9ETFJ0g3AR8smCyLiAUndwBZgGvBIRJxS5308slpHG/XImga1NBHoLsvPls9nJPPXl8v1Av2/MTFP0qTGumxmtYzkaPCngf4vFfpe+Twrmb+tRrkbqHmgSdJSSatG0A+zjtJUWCVdBnypnPwNcH3/rKEWSco197cjYnlEzG+mH2adqOGwSrocuLVc5m7gPRGxt5y9OWk6PSlPK597Kf5+NbMRaiiskj4AfL9s/2tgcURsT5o8lJTnlMt0A68v6x6LiN2Y2YjV/cI0SR8EbqEI6p3AhRHx6oBmK6geDV4maSlwJdWRdUVrurt/NfO9wd0zZ1bKvVsG7zSM9nuDn71mYaXu6K/fM2xf+hbV/p3TrrsfHnY5y0sjI+vnk3aLgR5JkTyOj4h7qR5supBit/ir5fQjwLdb2WmzTtTKa4OvAq6hOHWzB3geuAk4t97VS2ZWn3+f1axN+PdZzQ4QDqtZJhxWs0w4rGaZcFjNMuGwmmXCYTXLhMNqlgmH1SwTDqtZJhxWs0w4rGaZcFjNMuGwmmXCYTXLRN2vdRlvexafWSlPvPPBSrnrlLkA9K15Yr+874TjZ1fK+55+ZtD8el/rYq2l+ScDEKseHeeejB+PrGaZcFjNMuGvdbFR635T9UcGe9euG8ee5M1f62J2gGj7A0zW/jyajg2PrGaZcFjNMuGwmmXCYTXLhMNqlgmH1SwTDqtZJhxWs0w4rGaZcFjNMuHLDW3UfCH/2PDIapYJh9UsEw6rWSYcVrNMOKxmmfDRYBs1HwEeGx5ZzTLhsJplwmE1y0TTYZV0vqRIHosGzL9C0u8lvSpps6TbJB3bui6bdaamwirpIOCGYeZfC9wCnAJMAmYBS4B7JB05in6adbxmR9ZPAHOBXQNnSDoO+Jdy8n7gKODScvoY4Asj66KZQRNhlfQ64HPAn4F/q9HkfcBBZflbEfFCRNwGPF7WXSTJfyObjVAz4fkmMBW4BthaY/4ZSXl9jfIM4ISmemdmFQ2FVdI5wEXAPcAPhmg2KylvG6J8eI3XXippVSP9MOtkda9gkjQB+A7QC3wkIkIa9Js5ADUrB9QP+hWsiFgOLE9/mGq8TDj2mOpEd/X/sZH8PuuEE46rlPtmHFIpd23dURaqrx87q4cAek4vlkt/i/Z/nnu4Un7LFz4EwKzl9w69IgM8e83CSnnahr5K+eCXewHYfOpBNedPX3EfANsuPqvm607oKTbZtPXVHa29MydXyjtmH1wpv+bRok2939NN+zpzfW+lPGXl/cMuN1rp7wDvmzz8GNbfl/Rz2TO1+s883TZDfXYD9X/Ww2nkcsP3AicD/wUg6TQgPbJ7oqQXgM3peyflaUk5bWNmTWhkN3hq+XwBsLp8XJXMv5XidM1DSd2cGuVXgKdG1k0zq/v7rJI+QBHI4dwFXAb8geKI8H3AhcDfAreVbW6OiH8c5n38+6zW0Ub9+6wR8e8RofQBfDFp8tcRcW5EPAN8qaw7C3iealCfxedZzUalpec9I+I64EpgDbAbeAlYASyMiBda+V5mnabubvBY8W6wdbpR7wabWXtwWM0y4bCaZcJhNcuEw2qWCYfVLBMOq1kmHFazTDisZplwWM0y4bCaZcJhNcuEw2qWCYfVLBMOq1kmHFazTDisZplwWM0y4bCaZcJhNcuEw2qWCYfVLBMOq1kmHFazTDisZplwWM0y4bCaZcJhNcuEw2qWCYfVLBMOq1kmHFazTDisZplwWM0y4bCaZcJhNcuEw2qWCYfVLBMOq1kmHFazTDQVVkmXSrpH0nZJOyU9Ien6ZH63pKslrZO0W9Jzkm6UdGjru27WWRoOq6TvAD8EzgamAlOANwIXJ81uBq4H5gATgaOADwG/lTS5RX0260gNhVXSBcBHysk7KMI4BTgZ+ErZ5mzgirLNSuAw4DPl9JuBj7emy2adqdGR9WPl89PAJRHxfxHRExFrI+LGcl46wn4tIl4EvgFsrzHfzJpUN6ySuoG3lZMbgZWStkp6WdKPJB1VzjsjWWw9QET0Ak+WdfMkTWpRv806zoQG2ryWYpcXYNGAeRcD8yWdDsxK6rfVKHcDhwLPj6CfZh2vkd3ggwZMXw5MpziYBMXfr0sADbF8Wh+DZkpLJa1qoB9mHa2RsG6hGrItEfHDiNgO3Ji0ORXYnExPT8rTyufe8rX+QkQsj4j5jXfZrDPVDWtE7ALW9U8O0awHeCiZngOVv3dfX9Y9FhG7R9hPs47X6NHg28vnQyVdJmkq8OFk/l3AimR6maRZwNVUR9Z0vpk1SRFDDZZJoyKcDwAn1Zj9C+AdERGSbqF6rjX1CLAgInqGeY9KR87X++r2yexA88v4SaUcEYOOATU0skbEDuAcYDmwCdhLcUrmOuDdUU38VcA1FKdu9lAc+b0JOHe4oJpZfQ2NrGPBI6t1unojayPnWcfcnsVnMvHOByvlfv11AF2nzAWgb80Tlbp9f/OWSrlvYvM3FE1Zm5wC7q4uv2vuEcV77umr1E1a/WSl3LPgxEGvlfa1nu4jDq+Uezf9ueHlRmLbxWdVytNX3NeS1+x+0xsr5d6162q2qbW9mqH5JwMQqx5teJlnr1k47Pyjv35PpZz+O9s3udj2U1beP+zy++OzHI5vkTPLhHeDExOOPaY6kYys+55+ZlDb7pkzK+XeLYNOH5s1rSUHmMxs/DmsZploywNM42Xfho0Nt6236zvhhOOqr/vUn6r1x88uCl3V/ydj567q6+7nA0yWL4+sZplwWM0y4d3g/STd9f2L+hpHls0a4ZHVLBMeWUfI51ltrHlkNcuEw2qWibbfDU4vsO75WHV3c+ayouvjdSF/d3Ihf9rHymutq54v7ZtxSKXctXVHWah9nrXn9OL8bDM3ArSTvkWnVcpddz/c9PLpxffphfb1LuR/5b8H30yx467Da7SsmnpOdRvNeOcfBs3fdeGCmsvVu8B/f/HIapYJh9UsE77rxqxN+K4bswNE2x9gOtD4Qn4bKY+sZplwWM0y4d3gMeYL+W2kPLKaZcJhNcuEw2qWCYfVLBM+wDRCvp+1OaP9Rn7zyGqWDYfVLBPeDR4h7/o2x7u/o+eR1SwTDqtZJrwbPITK3TH4V+SsPXhkNcuEw2qWCe8GD6He3THe9bWx5pHVLBMeWUfIB5hsrHlkNcuEw2qWCe8Gj5B3fW2sNTyySrpM0r2SXpS0U9I6SV+VNDNp0y3p6nLebknPSbpR0qH7p/tmnaOhkVXS1cD1A6rnAMuA84CzyrqbgSuSNkcBHwIWSVoQET2j665Z52p0ZL2kfO4FzgFmAQ+UdQskzZN0NtWgrgQOAz5TTr8Z+Pjou1vVdcrcyg3NrTLh2GMqj3q6Z86sPMzGQqNh7S2fX4iI30XES8AvkvmTgYuT6a9FxIvAN4DtZV0638ya1OgBpuXATcBRkt4OrAX+rpz3HPAocEbSfj1ARPRKehI4FZgnaVJE7G5Fx/fH/ZH7NmxsuK0PMNlYa2hkjYjvAv8ECLgLeBF4K7AaeFcZwFnJIttqlLuBQQeaJC2VtKr5rpt1mIio+6DYhd0LxIDHc8DlZZv1SX1Xsuz/JvVHDvMeA1/bDz869lErI3VHVkldwA0Uu8zPAvOA6cAPKI723ippPrA5WWx6Up5WPvcC3nc0G6FGdoMPp7r7+ruIeDwitgMryjpRnL55KFlmDhTnXYHXl3WPtervVbNO1MgBpi3Aq8DBwNslzaUYYdOju1uBu4GPltPLJC0FrqQ6sq5gGOkvPUtaFRHzG1qDTByI6wRer7FUd2QtR8ObysmjgccpDhpdXtZtAn4aEfcC3yvrLqTYLf5qOf0I8O0W9dmsIzV6nvVqiqPBq4FdwD6K0fU/gIXleVeAq4BrKA427QGepwj6uU1evbS8iba5OBDXCbxeY0blkVgza3O+Rc4sE20TVkmzJd0mabOkHkm/l3RF/SXHX3lH0kpJT0naJWmTpF9JOm9Au2zvSpJ0vqRIHosGzL+i3GavltvwNknHjld/GyHpUkn3SNpe3kn2hKTrk/nttb0auShifz+AI4GN1D5B/Nnx7l8D/X9iiL4HcFHS7pYh2qwBJo/3egyzfgdRHFhM+7womX/tEOu1gWEuhBnndfrOEH3e2K7ba9w/tPJDuTn5IC4pw3tfOb0HmD3efazT//uAzwLHUZyq+nKyPmvLNmcndT+luDxzWVK3bLzXY5j1++eyjzsHhrVc5z1l3X3ltrskaffd8e5/jfW5IOnffwJ/RXEzypuAD7fr9mqHD66L4jxtUFw40V///uSD+eR497POOkwdMC3glbLvu8u6G5L1eWtZ101xGiyANeO9HkOs2+so7pzaBPxrjbB+Kql7f7LcY2XdVpLLT9vhAfy87NtTwMQh2rTd9mqHv1nfAMwoy+uT+rR8+th1p3kRsWNA1USKDQvFKS4Y4q4k4Mmybp6kSfutkyP3TWAqxSm5rTXmD1qvAeUZwAn7p2vNK6+qe1s5uRFYKWmrpJcl/UjSUeW8ttte7RDWenfrQHHJY04+DRxSlvsvFBnRXUnjSdI5wEXAPRTXgteS2/Z7LTClLC8C3knxH8pMiqvyfitpCm24vdohrGqgPpuTwZIuA75UTv6G6tfhZLWekiZQHITpBT4S5X5graYN1LfNelEcLEtdTnHjyc3l9BxgCW24Xu0Q1np36wxs07YkXQ7cSvG53g28JyL2lrNzuyvpvcDJwM8AJJ1GcfCo34mSTiS/7beFasi2RMQPo7gx5cakzam04fZqh7D+keJgDJR369Qorx677oyMpA8A36f4TH8NLC7/EfTL7a6kqeXzBRSf/2qKy0n73UpxamPQeg0ov0JxIKctRMQuYF3/5BDNemjD7TXuYY2IPuDH5eRJkpZIOgL4ZFm3F7hjXDrXIEkfpPjbtAu4k+LbM3YOaJbedbRM0iyKa64buiupjd1BsY0APiHpSElLgJPKutvLbdxObi+fDy0vaJkKfDiZfxftuL3G+zB6+adQ7hdFPD1E3/sfx5ft2uok+wjW8wtJnxu5KGIjbXhRBMVew2ND9PnnVK+Zb6vtNe4jK0BEvAAspPjf6iVgN8UHcmVEfHk8+9Zirborqa1ExHUU9y6vodh2L1Fsy4Xltm0rUZxqO4fizppNFHsGTwLXAe+OMqm02fbyXTdmmWiLkdXM6nNYzTLhsJplwmE1y4TDapYJh9UsEw6rWSYcVrNMOKxmmfh/jpc/Tz78mgIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "b_real = np.zeros((MAX_ID,))\n",
    "b_real[thai_metagenome_ID] = 1\n",
    "[m2b, b2m] = Ain_out(b_real, net)\n",
    "\n",
    "plt.figure()\n",
    "plt.imshow(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)])\n",
    "\n",
    "#print(np.sum(m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)], 0))\n",
    "#print(np.sum(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)], 0))\n",
    "#print(np.sum(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)], 1))\n",
    "\n",
    "print('Consumption connectance is: ', np.sum(m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0) / (thai_metagenome_ID.shape[0]*thai_metabolome_ID.shape[0]))\n",
    "print('Production connectance is: ', np.sum(b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0) / (thai_metagenome_ID.shape[0]*thai_metabolome_ID.shape[0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "m2b_ori = (m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "b2m_ori = (b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "net_ori = net.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4248 190 177\n"
     ]
    }
   ],
   "source": [
    "print(len(net_ori), len(net_added_consumption), len(net_added_production))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "net = pd.concat([net_ori, net_added_consumption, net_added_production])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gradient descent over adding links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6624662145232207 0.0\n",
      "0.6624662145232207 0.0\n",
      "0.6624662145381219 1.4901161193847656e-08\n",
      "0.6624662145381219 1.4901161193847656e-08\n",
      "0.6624662145381219 1.4901161193847656e-08\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-249-bcb7cbfa1c63>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     94\u001b[0m \u001b[0mx0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx_ori\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     95\u001b[0m \u001b[0mbnds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mmax_links\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 96\u001b[0;31m \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mminimize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'SLSQP'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbounds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbnds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moptions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'disp'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'maxiter'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m1000\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtol\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1e-3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     97\u001b[0m \u001b[0;31m#res = minimize(fun, x0, method='Nelder-Mead', bounds=bnds, options={'disp': True})\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m \u001b[0;31m#res = minimize(fun, x0, method='TNC', bounds=bnds, options={'disp': True, 'maxiter': 1000})\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/_minimize.py\u001b[0m in \u001b[0;36mminimize\u001b[0;34m(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)\u001b[0m\n\u001b[1;32m    606\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mmeth\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'slsqp'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m         return _minimize_slsqp(fun, x0, args, jac, bounds,\n\u001b[0;32m--> 608\u001b[0;31m                                constraints, callback=callback, **options)\n\u001b[0m\u001b[1;32m    609\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mmeth\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'trust-constr'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    610\u001b[0m         return _minimize_trustregion_constr(fun, x0, args, jac, hess, hessp,\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/slsqp.py\u001b[0m in \u001b[0;36m_minimize_slsqp\u001b[0;34m(func, x0, args, jac, bounds, constraints, maxiter, ftol, iprint, disp, eps, callback, **unknown_options)\u001b[0m\n\u001b[1;32m    421\u001b[0m             \u001b[0;31m# Compute the derivatives of the objective function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    422\u001b[0m             \u001b[0;31m# For some reason SLSQP wants g dimensioned to n+1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 423\u001b[0;31m             \u001b[0mg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfprime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0.0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    424\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    425\u001b[0m             \u001b[0;31m# Compute the normals of the constraints\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py\u001b[0m in \u001b[0;36mfunction_wrapper\u001b[0;34m(*wrapper_args)\u001b[0m\n\u001b[1;32m    324\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mwrapper_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m         \u001b[0mncalls\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwrapper_args\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mncalls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/slsqp.py\u001b[0m in \u001b[0;36mapprox_jacobian\u001b[0;34m(x, func, epsilon, *args)\u001b[0m\n\u001b[1;32m     61\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m         \u001b[0mdx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mepsilon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 63\u001b[0;31m         \u001b[0mjac\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx0\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mdx\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mf0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mepsilon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     64\u001b[0m         \u001b[0mdx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0.0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/scipy/optimize/optimize.py\u001b[0m in \u001b[0;36mfunction_wrapper\u001b[0;34m(*wrapper_args)\u001b[0m\n\u001b[1;32m    324\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mwrapper_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    325\u001b[0m         \u001b[0mncalls\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 326\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwrapper_args\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    328\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mncalls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunction_wrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-249-bcb7cbfa1c63>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m     90\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mpred_error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     91\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 92\u001b[0;31m \u001b[0mfun\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpred_error_addingLinks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm2b_ori\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb2m_ori\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx_ori\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     93\u001b[0m \u001b[0mmax_links\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mthai_metabolome_ID\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mthai_metagenome_ID\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     94\u001b[0m \u001b[0mx0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx_ori\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-249-bcb7cbfa1c63>\u001b[0m in \u001b[0;36mpred_error_addingLinks\u001b[0;34m(x, m2b_ori, b2m_ori, x_ori)\u001b[0m\n\u001b[1;32m     48\u001b[0m         \u001b[0mm2b\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb2m\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAin_out\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb_real\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     49\u001b[0m         \u001b[0;31m#m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(f, m2b, b2m, numLevels_max)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 50\u001b[0;31m         \u001b[0mm2b_total\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm2m_total\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm2m_layer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mm2b_multiple_levels\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi_nonzero_microbes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi_nonzero_metabolites\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mm2b\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb2m\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumLevels_max\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     51\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m         \u001b[0;31m######## As long as the optimized nutrient intake is found by using the above optimization solver, the\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-247-7d30942ee1e7>\u001b[0m in \u001b[0;36mm2b_multiple_levels\u001b[0;34m(i_nonzero_microbes, i_nonzero_metabolites, f, m2b, b2m, numLevels_max)\u001b[0m\n\u001b[1;32m    106\u001b[0m         \u001b[0;31m# m2b_total is a series made of s_step of each layer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    107\u001b[0m         \u001b[0mm2b_total\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mm2b_total\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mf_mul\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mii\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0ms_step_ii\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 108\u001b[0;31m         \u001b[0mm2m_layer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mii\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf_mul\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mii\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0ms_step_ii\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    109\u001b[0m         \u001b[0ms_step_ii\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marray_mult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi_nonzero_metabolites\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi_nonzero_metabolites\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi_nonzero_metabolites\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi_nonzero_metabolites\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms_step_ii\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms_step\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    110\u001b[0m     \u001b[0mm2m_total\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mm2b_total\u001b[0m\u001b[0;34m;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "m2b_ori = (m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "b2m_ori = (b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "x_ori = np.concatenate([m2b_ori.flatten(), b2m_ori.flatten()])\n",
    "net_ori = net.copy()\n",
    "\n",
    "def pred_error_addingLinks(x, m2b_ori, b2m_ori, x_ori):\n",
    "    '''\n",
    "    pred_error_addingLinks is a function used to compute the logarithmic error between experimentally measured\n",
    "    metagenome and predicted metagenome computed from the model for a certain nutrient intake. It relies on \n",
    "    (1) x: the nutrient intake, (2) i_intake: IDs of the nutrient intake, (3) m2b_total: a conversion matrix \n",
    "    from thenutrient intake to the total biomass, and (4) b_real: experimentally measured metagenome. The \n",
    "    first three is used to compute the metagenome predicted by the model \"ba_pred\" and compare it with the \n",
    "    experimentally measured metagenome \"b_real\".\n",
    "    '''\n",
    "    max_links = thai_metabolome_ID.shape[0]*thai_metagenome_ID.shape[0]\n",
    "    \n",
    "    ######## Convert x to net structure:\n",
    "    thres = 1e-10\n",
    "    x = x - x_ori\n",
    "    m2b_added = x[:max_links].reshape((thai_metabolome_ID.shape[0], thai_metagenome_ID.shape[0]))\n",
    "    a = thai_metabolome_ID.iloc[np.where(m2b_added >= thres)[0]]\n",
    "    b = thai_metagenome_ID.iloc[np.where(m2b_added >= thres)[1]]\n",
    "    c = [2] * np.where(m2b_added >= thres)[1].shape[0]\n",
    "    net_added_consumption = pd.DataFrame({net_ori.columns[0]:list(a.values), net_ori.columns[1]:list(b.values), net_ori.columns[2]:c})\n",
    "\n",
    "    b2m_added = x[max_links:].reshape((thai_metabolome_ID.shape[0], thai_metagenome_ID.shape[0]))\n",
    "    a = thai_metabolome_ID.iloc[np.where(b2m_added >= thres)[0]]\n",
    "    b = thai_metagenome_ID.iloc[np.where(b2m_added >= thres)[1]]\n",
    "    c = [3] * np.where(b2m_added >= thres)[1].shape[0]\n",
    "    net_added_production = pd.DataFrame({net_ori.columns[0]:list(a.values), net_ori.columns[1]:list(b.values), net_ori.columns[2]:c})\n",
    "    \n",
    "    net = pd.concat([net_ori, net_added_consumption, net_added_production])\n",
    "    \n",
    "    f_byproduct = 0.9\n",
    "    f = f_byproduct * np.ones((MAX_ID,1));\n",
    "    f[i_selfish] = 0.0;  # The byproduct/leakage fraction f for microbes that don't generate byproducts is set as 0.\n",
    "\n",
    "    numLevels_max = 4\n",
    "\n",
    "    corr_list_aveDiet = np.zeros((4,2))\n",
    "    for pa in range(4):\n",
    "        #print(pa)\n",
    "        ######## Select an ID of a individual and pull out its experimentally measured metagenome:\n",
    "        b_real = np.zeros((MAX_ID,))\n",
    "        b_real[thai_metagenome_ID] = thai_metagenome.iloc[:,pa] / np.sum(thai_metagenome.iloc[:,pa])\n",
    "\n",
    "        ######## Compute matrices involving the metabolite consumption and generation:\n",
    "        m2b, b2m = Ain_out(b_real, net)\n",
    "        #m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(f, m2b, b2m, numLevels_max)\n",
    "        m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(i_nonzero_microbes, i_nonzero_metabolites, f, m2b, b2m, numLevels_max)\n",
    "\n",
    "        ######## As long as the optimized nutrient intake is found by using the above optimization solver, the\n",
    "        # optimized nutrient intake \"res.x\" is used to generate the predict metagenome and metabolome. They are\n",
    "        # visually and statistically compared to the experimentally measured metagenome and metabolome.\n",
    "        x_full = np.zeros((MAX_ID,));\n",
    "        x_full[i_intake] = average_diet.copy();\n",
    "        ba_pred = np.dot(m2b_total, x_full)\n",
    "\n",
    "        met_levels, met_leftover_levels = calc_metabolome(average_diet.copy(), m2m_layer, numLevels_max, i_intake)\n",
    "        metabolome_measured = np.zeros((MAX_ID,))\n",
    "        metabolome_measured[thai_metabolome_ID.values] = thai_metabolome.iloc[:,pa]\n",
    "        metabolome_pred = met_levels[:, -1] + met_leftover_levels.sum(1) #np.dot(m2m_total, x_full)\n",
    "        i_common = np.where(metabolome_measured * metabolome_pred > 1e-5)[0]\n",
    "        metabolome_pred_common = metabolome_pred[i_common] / np.sum(metabolome_pred[i_common])\n",
    "        metabolome_measured_common = metabolome_measured[i_common] / np.sum(metabolome_measured[i_common])\n",
    "\n",
    "        corr_list_aveDiet[pa, 0] = pearsonr(ba_pred[b_real>0], b_real[b_real>0])[0]\n",
    "        corr_list_aveDiet[pa, 1] = pearsonr(metabolome_pred_common, metabolome_measured_common)[0]\n",
    "        '''\n",
    "        print('-------------------------------------------------------------------------------------------------------------')\n",
    "        print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:')\n",
    "        print(pearsonr(ba_pred[b_real>0], b_real[b_real>0]))\n",
    "        print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:')\n",
    "        print(pearsonr(metabolome_pred_common, metabolome_measured_common))\n",
    "        '''  \n",
    "        \n",
    "    x_full = np.zeros((MAX_ID,));\n",
    "    x_full[i_intake] = average_diet.copy();\n",
    "    \n",
    "    ba_pred = np.dot(m2b_total, x_full)\n",
    "    #pred_error1 = (np.log10(ba_pred + 1e-6) - np.log10(b_real +1e-6)) / np.log10(b_real +1e-6);\n",
    "    #pred_error2 = (np.log10(metabolome_pred_common + 1e-6) - np.log10(metabolome_measured_common + 1e-6)) / np.log10(metabolome_measured_common + 1e-6);\n",
    "    pred_error1 = np.mean(corr_list_aveDiet[:,0])\n",
    "    pred_error2 = np.mean(corr_list_aveDiet[:,1])\n",
    "    pred_error3 = np.sum(np.abs(x))\n",
    "    hyper_reg = 0.001\n",
    "    pred_error = 1 - pred_error1 + 1 - pred_error2 + hyper_reg * pred_error3\n",
    "    print(pred_error, pred_error3)\n",
    "    \n",
    "    return pred_error\n",
    "\n",
    "fun = lambda x: pred_error_addingLinks(x, m2b_ori, b2m_ori, x_ori)\n",
    "max_links = thai_metabolome_ID.shape[0]*thai_metagenome_ID.shape[0]\n",
    "x0 = x_ori.copy()\n",
    "bnds = ((0, 100), ) * max_links * 2\n",
    "res = minimize(fun, x0, method='SLSQP', bounds=bnds, options={'disp': True, 'maxiter': 1000}, tol=1e-3)\n",
    "#res = minimize(fun, x0, method='Nelder-Mead', bounds=bnds, options={'disp': True})\n",
    "#res = minimize(fun, x0, method='TNC', bounds=bnds, options={'disp': True, 'maxiter': 1000})\n",
    "print(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monte Carlo over adding links (using the average diet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.6665375599602643 0\n",
      "1.6722135064658097 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6730344225337872 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6782399216523276 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6789812774891735 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765498263458898 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.678826792709223 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.696270305314725 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6708726772531592 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6854009229746467 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765315608655238 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6754213445032253 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.67815005563318 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6665375599602643 0\n",
      "accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6773811635391358 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6835075054894222 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6747179086010229 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6737361838164415 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6765375599602643 1\n",
      "not accepted, error is 1.6665375599602643\n",
      "1.6624434703436688 1\n",
      "accepted, error is 1.6624434703436688\n",
      "1.7115060453704376 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6624434703436688 1\n",
      "accepted, error is 1.6624434703436688\n",
      "1.7083175684704959 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6756980740493987 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.67244438663905 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6624434703436688 1\n",
      "accepted, error is 1.6624434703436688\n",
      "1.670777170513695 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.683590053717245 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6670622915120137 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6724434703436688 2\n",
      "not accepted, error is 1.6624434703436688\n",
      "1.6178859916132944 2\n",
      "accepted, error is 1.6178859916132944\n",
      "1.6601458266989584 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6278859916132944 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6274318823013136 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6278869079088412 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6423186865107728 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6299328068836718 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6282872861192978 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6278859916132944 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.663580869993981 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6178859916132944 2\n",
      "accepted, error is 1.6178859916132944\n",
      "1.6278700006450666 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.621337969891491 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6278859916132944 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6360476992170936 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6551162652471754 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.630120177062166 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.6278859916132944 3\n",
      "not accepted, error is 1.6178859916132944\n",
      "1.618785660921563 3\n",
      "accepted, error is 1.618785660921563\n",
      "1.6306455483233158 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.631182463582517 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.6532835982987981 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.6288907469593092 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.6287520593689404 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.6265801001817117 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.618785660921563 3\n",
      "accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.6284535861241747 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.618785660921563 3\n",
      "accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.6287487527338702 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.634417510523336 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.628785660921563 4\n",
      "not accepted, error is 1.618785660921563\n",
      "1.617917864841203 4\n",
      "accepted, error is 1.617917864841203\n",
      "1.679149803327597 5\n",
      "not accepted, error is 1.617917864841203\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "m2b_ori = (m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "b2m_ori = (b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "x_ori = np.concatenate([m2b_ori.flatten(), b2m_ori.flatten()])\n",
    "net_ori = net.copy()\n",
    "\n",
    "def pred_error_addingLinks(x, m2b_ori, b2m_ori, x_ori):\n",
    "    '''\n",
    "    pred_error_addingLinks is a function used to compute the logarithmic error between experimentally measured\n",
    "    metagenome and predicted metagenome computed from the model for a certain nutrient intake. It relies on \n",
    "    (1) x: the nutrient intake, (2) i_intake: IDs of the nutrient intake, (3) m2b_total: a conversion matrix \n",
    "    from thenutrient intake to the total biomass, and (4) b_real: experimentally measured metagenome. The \n",
    "    first three is used to compute the metagenome predicted by the model \"ba_pred\" and compare it with the \n",
    "    experimentally measured metagenome \"b_real\".\n",
    "    '''\n",
    "    max_links = thai_metabolome_ID.shape[0]*thai_metagenome_ID.shape[0]\n",
    "    \n",
    "    ######## Convert x to net structure:\n",
    "    thres = 0.1\n",
    "    x = x - x_ori\n",
    "    m2b_added = x[:max_links].reshape((thai_metabolome_ID.shape[0], thai_metagenome_ID.shape[0]))\n",
    "    a = thai_metabolome_ID.iloc[np.where(m2b_added >= thres)[0]]\n",
    "    b = thai_metagenome_ID.iloc[np.where(m2b_added >= thres)[1]]\n",
    "    c = [2] * np.where(m2b_added >= thres)[1].shape[0]\n",
    "    net_added_consumption = pd.DataFrame({net_ori.columns[0]:list(a.values), net_ori.columns[1]:list(b.values), net_ori.columns[2]:c})\n",
    "\n",
    "    b2m_added = x[max_links:].reshape((thai_metabolome_ID.shape[0], thai_metagenome_ID.shape[0]))\n",
    "    a = thai_metabolome_ID.iloc[np.where(b2m_added >= thres)[0]]\n",
    "    b = thai_metagenome_ID.iloc[np.where(b2m_added >= thres)[1]]\n",
    "    c = [3] * np.where(b2m_added >= thres)[1].shape[0]\n",
    "    net_added_production = pd.DataFrame({net_ori.columns[0]:list(a.values), net_ori.columns[1]:list(b.values), net_ori.columns[2]:c})\n",
    "    \n",
    "    net = pd.concat([net_ori, net_added_consumption, net_added_production])\n",
    "    \n",
    "    f_byproduct = 0.9\n",
    "    f = f_byproduct * np.ones((MAX_ID,1));\n",
    "    f[i_selfish] = 0.0;  # The byproduct/leakage fraction f for microbes that don't generate byproducts is set as 0.\n",
    "\n",
    "    numLevels_max = 4\n",
    "\n",
    "    corr_list_aveDiet = np.zeros((41,2))\n",
    "    log_list_aveDiet = np.zeros((41,2))\n",
    "    for pa in range(41):\n",
    "        #print(pa)\n",
    "        ######## Select an ID of a individual and pull out its experimentally measured metagenome:\n",
    "        b_real = np.zeros((MAX_ID,))\n",
    "        b_real[thai_metagenome_ID] = thai_metagenome.iloc[:,pa] / np.sum(thai_metagenome.iloc[:,pa])\n",
    "\n",
    "        ######## Compute matrices involving the metabolite consumption and generation:\n",
    "        m2b, b2m = Ain_out(b_real, net)\n",
    "        #m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(f, m2b, b2m, numLevels_max)\n",
    "        m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(i_nonzero_microbes, i_nonzero_metabolites, f, m2b, b2m, numLevels_max)\n",
    "\n",
    "        ######## As long as the optimized nutrient intake is found by using the above optimization solver, the\n",
    "        # optimized nutrient intake \"res.x\" is used to generate the predict metagenome and metabolome. They are\n",
    "        # visually and statistically compared to the experimentally measured metagenome and metabolome.\n",
    "        x_full = np.zeros((MAX_ID,));\n",
    "        x_full[i_intake] = average_diet.copy();\n",
    "        ba_pred = np.dot(m2b_total, x_full)\n",
    "\n",
    "        met_levels, met_leftover_levels = calc_metabolome(average_diet.copy(), m2m_layer, numLevels_max, i_intake)\n",
    "        metabolome_measured = np.zeros((MAX_ID,))\n",
    "        metabolome_measured[thai_metabolome_ID.values] = thai_metabolome.iloc[:,pa]\n",
    "        metabolome_pred = met_levels[:, -1] + met_leftover_levels.sum(1) #np.dot(m2m_total, x_full)\n",
    "        i_common = np.where(metabolome_measured * metabolome_pred > 1e-5)[0]\n",
    "        metabolome_pred_common = metabolome_pred[i_common] / np.sum(metabolome_pred[i_common])\n",
    "        metabolome_measured_common = metabolome_measured[i_common] / np.sum(metabolome_measured[i_common])\n",
    "\n",
    "        corr_list_aveDiet[pa, 0] = pearsonr(ba_pred[b_real>0], b_real[b_real>0])[0]\n",
    "        corr_list_aveDiet[pa, 1] = pearsonr(metabolome_pred_common, metabolome_measured_common)[0]\n",
    "        log_list_aveDiet[pa, 0] = np.mean(np.abs(np.log10(ba_pred[b_real>0]+1e-6) - np.log10(b_real[b_real>0]+1e-6)))\n",
    "        log_list_aveDiet[pa, 1] = np.mean(np.abs(np.log10(metabolome_pred_common+1e-6) - np.log10(metabolome_measured_common+1e-6)))\n",
    "        '''\n",
    "        print('-------------------------------------------------------------------------------------------------------------')\n",
    "        print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metagonome:')\n",
    "        print(pearsonr(ba_pred[b_real>0], b_real[b_real>0]))\n",
    "        print('(Correlation coefficient, P-value) of the correlation between predicted and experimentally measured metabolome:')\n",
    "        print(pearsonr(metabolome_pred_common, metabolome_measured_common))\n",
    "        '''  \n",
    "        \n",
    "    x_full = np.zeros((MAX_ID,));\n",
    "    x_full[i_intake] = average_diet.copy();\n",
    "    \n",
    "    ba_pred = np.dot(m2b_total, x_full)\n",
    "    #pred_error1 = (np.log10(ba_pred + 1e-6) - np.log10(b_real +1e-6)) / np.log10(b_real +1e-6);\n",
    "    #pred_error2 = (np.log10(metabolome_pred_common + 1e-6) - np.log10(metabolome_measured_common + 1e-6)) / np.log10(metabolome_measured_common + 1e-6);\n",
    "    #pred_error1 = np.mean(corr_list_aveDiet[:,0])\n",
    "    #pred_error2 = np.mean(corr_list_aveDiet[:,1])\n",
    "    #pred_error3 = np.sum(np.abs(x))\n",
    "    #hyper_reg = 0.01\n",
    "    #pred_error = 1 - pred_error1 + 1 - pred_error2 + hyper_reg * pred_error3\n",
    "    \n",
    "    pred_error1 = np.mean(log_list_aveDiet[:,0])\n",
    "    pred_error2 = np.mean(log_list_aveDiet[:,1])\n",
    "    pred_error3 = np.sum(np.abs(x))\n",
    "    hyper_reg = 0.01\n",
    "    pred_error = pred_error1 + pred_error2 + hyper_reg * pred_error3\n",
    "    print(pred_error, pred_error3)\n",
    "    \n",
    "    return pred_error\n",
    "\n",
    "fun = lambda x: pred_error_addingLinks(x, m2b_ori, b2m_ori, x_ori)\n",
    "max_links = thai_metabolome_ID.shape[0]*thai_metagenome_ID.shape[0]\n",
    "x = x_ori.copy()\n",
    "error_before = fun(x)\n",
    "for i in range(100):\n",
    "    i_x = random.randint(0,max_links*2)\n",
    "    x[i_x] = 1\n",
    "    error_after = fun(x)\n",
    "    if np.random.uniform(0,1,1)[0] > np.exp(1000*(error_before - error_after)):  ## not accepted\n",
    "        x[i_x] = 0\n",
    "        print('not accepted, error is', error_before)\n",
    "    else:  ## accepted   \n",
    "        error_before = error_after\n",
    "        print('accepted, error is', error_before)\n",
    "    #print(error_after / error_before)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7073242655976671"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#pearsonr(ba_pred[b_real>0], b_real[b_real>0])[0]\n",
    "\n",
    "np.mean(np.abs(np.log10(ba_pred[b_real>0]+1e-5) - np.log10(b_real[b_real>0]+1e-5)))\n",
    "np.mean(np.abs(np.log10(metabolome_pred_common+1e-6) - np.log10(metabolome_measured_common+1e-6)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monte Carlo over adding links (using the optimized diet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "m2b_ori = (m2b[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "b2m_ori = (b2m[np.ix_(thai_metabolome_ID,thai_metagenome_ID)]!=0).astype(int).copy()\n",
    "x_ori = np.concatenate([m2b_ori.flatten(), b2m_ori.flatten()])\n",
    "net_ori = net.copy()\n",
    "\n",
    "def pred_error_addingLinks(x, m2b_ori, b2m_ori, x_ori):\n",
    "    '''\n",
    "    pred_error_addingLinks is a function used to compute the logarithmic error between experimentally measured\n",
    "    metagenome and predicted metagenome computed from the model for a certain nutrient intake. It relies on \n",
    "    (1) x: the nutrient intake, (2) i_intake: IDs of the nutrient intake, (3) m2b_total: a conversion matrix \n",
    "    from thenutrient intake to the total biomass, and (4) b_real: experimentally measured metagenome. The \n",
    "    first three is used to compute the metagenome predicted by the model \"ba_pred\" and compare it with the \n",
    "    experimentally measured metagenome \"b_real\".\n",
    "    '''\n",
    "    max_links = thai_metabolome_ID.shape[0]*thai_metagenome_ID.shape[0]\n",
    "    \n",
    "    ######## Convert x to net structure:\n",
    "    thres = 0.1\n",
    "    x = x - x_ori\n",
    "    m2b_added = x[:max_links].reshape((thai_metabolome_ID.shape[0], thai_metagenome_ID.shape[0]))\n",
    "    a = thai_metabolome_ID.iloc[np.where(m2b_added >= thres)[0]]\n",
    "    b = thai_metagenome_ID.iloc[np.where(m2b_added >= thres)[1]]\n",
    "    c = [2] * np.where(m2b_added >= thres)[1].shape[0]\n",
    "    net_added_consumption = pd.DataFrame({net_ori.columns[0]:list(a.values), net_ori.columns[1]:list(b.values), net_ori.columns[2]:c})\n",
    "\n",
    "    b2m_added = x[max_links:].reshape((thai_metabolome_ID.shape[0], thai_metagenome_ID.shape[0]))\n",
    "    a = thai_metabolome_ID.iloc[np.where(b2m_added >= thres)[0]]\n",
    "    b = thai_metagenome_ID.iloc[np.where(b2m_added >= thres)[1]]\n",
    "    c = [3] * np.where(b2m_added >= thres)[1].shape[0]\n",
    "    net_added_production = pd.DataFrame({net_ori.columns[0]:list(a.values), net_ori.columns[1]:list(b.values), net_ori.columns[2]:c})\n",
    "    \n",
    "    net = pd.concat([net_ori, net_added_consumption, net_added_production])\n",
    "    \n",
    "    f_byproduct = 0.9\n",
    "    f = f_byproduct * np.ones((MAX_ID,1));\n",
    "    f[i_selfish] = 0.0;  # The byproduct/leakage fraction f for microbes that don't generate byproducts is set as 0.\n",
    "\n",
    "    numLevels_max = 4\n",
    "\n",
    "    corr_list = np.zeros((4,2))\n",
    "    for pa in range(4):\n",
    "        ######## Select an ID of a individual and pull out its experimentally measured metagenome:\n",
    "        b_real = np.zeros((MAX_ID,))\n",
    "        b_real[thai_metagenome_ID] = thai_metagenome.iloc[:,pa] / np.sum(thai_metagenome.iloc[:,pa])\n",
    "\n",
    "        ######## Compute matrices involving the metabolite consumption and generation:\n",
    "        m2b, b2m = Ain_out(b_real, net)\n",
    "        m2b_total, m2m_total, m2m_layer = m2b_multiple_levels(f, m2b, b2m, numLevels_max)\n",
    "\n",
    "        ######## Assign diet\n",
    "        x_diet0 = np.ones((len(i_intake),)) / len(i_intake)\n",
    "\n",
    "        ######## The model is converted into an optimization problem where the nutrient intake is constantly changed to\n",
    "        # minimize the logarithmic error between experimentally measured metagenome and predicted metagenome computed \n",
    "        # from the model for a certain nutrient intake.\n",
    "        fun = lambda x_diet: pred_error(x_diet, b_real, m2b_total, i_intake)\n",
    "        bnds = ((0, 100), ) * len(x_diet0)\n",
    "        res = minimize(fun, x_diet0, method='SLSQP', bounds=bnds, options={'maxiter': 1000}, tol=1e-3)\n",
    "\n",
    "        ######## As long as the optimized nutrient intake is found by using the above optimization solver, the\n",
    "        # optimized nutrient intake \"res.x\" is used to generate the predict metagenome and metabolome. They are\n",
    "        # visually and statistically compared to the experimentally measured metagenome and metabolome.\n",
    "        x_full = np.zeros((MAX_ID,));\n",
    "        x_full[i_intake] = res.x;\n",
    "        ba_pred = np.dot(m2b_total, x_full)\n",
    "\n",
    "        met_levels, met_leftover_levels = calc_metabolome(res.x, m2m_layer, numLevels_max, i_intake)\n",
    "        metabolome_measured = np.zeros((MAX_ID,))\n",
    "        metabolome_measured[thai_metabolome_ID.values] = thai_metabolome.iloc[:,pa]\n",
    "        metabolome_pred = met_levels[:, -1] + met_leftover_levels.sum(1) #np.dot(m2m_total, x_full)\n",
    "        i_common = np.where(metabolome_measured * metabolome_pred > 1e-5)[0]\n",
    "        metabolome_pred_common = metabolome_pred[i_common] / np.sum(metabolome_pred[i_common])\n",
    "        metabolome_measured_common = metabolome_measured[i_common] / np.sum(metabolome_measured[i_common])\n",
    "\n",
    "        corr_list[pa, 0] = pearsonr(ba_pred[b_real>0], b_real[b_real>0])[0]\n",
    "        corr_list[pa, 1] = pearsonr(metabolome_pred_common, metabolome_measured_common)[0]\n",
    "        \n",
    "    x_full = np.zeros((MAX_ID,));\n",
    "    x_full[i_intake] = res.x;\n",
    "    \n",
    "    ba_pred = np.dot(m2b_total, x_full)\n",
    "    #pred_error1 = (np.log10(ba_pred + 1e-6) - np.log10(b_real +1e-6)) / np.log10(b_real +1e-6);\n",
    "    #pred_error2 = (np.log10(metabolome_pred_common + 1e-6) - np.log10(metabolome_measured_common + 1e-6)) / np.log10(metabolome_measured_common + 1e-6);\n",
    "    pred_error1 = np.mean(corr_list[:,0])\n",
    "    pred_error2 = np.mean(corr_list[:,1])\n",
    "    pred_error3 = np.sum(np.abs(x))\n",
    "    hyper_reg = 0.01\n",
    "    pred_error = 1 - pred_error1 + 1 - pred_error2 + hyper_reg * pred_error3\n",
    "    print(pred_error, pred_error3)\n",
    "    \n",
    "    return pred_error\n",
    "\n",
    "fun2 = lambda x: pred_error_addingLinks(x, m2b_ori, b2m_ori, x_ori)\n",
    "max_links = thai_metabolome_ID.shape[0]*thai_metagenome_ID.shape[0]\n",
    "x = x_ori.copy()\n",
    "error_before = fun2(x)\n",
    "for i in range(100):\n",
    "    i_x = random.randint(0,max_links*2)\n",
    "    x[i_x] = 1\n",
    "    error_after = fun2(x)\n",
    "    if np.random.uniform(0,1,1)[0] > np.exp(1000*(error_before - error_after)):  ## not accepted\n",
    "        x[i_x] = 0\n",
    "        print('not accepted, error is', error_after)\n",
    "    else:  ## accepted   \n",
    "        error_before = error_after\n",
    "        print('accepted, error is', error_after)\n",
    "    #print(error_after / error_before)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
